{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cardiovascular-homeless",
   "metadata": {},
   "source": [
    "-----\n",
    "# cartoGRAPHs\n",
    "precalculated large networks , e.g. the Interactome\n",
    "\n",
    "A Notebook to produce 2D and 3D network layouts from any Graph,\n",
    "including interactive visualization (html files) and export functions \n",
    "to import into the VRNetzer analytics platform by Pirch et al.\n",
    "\n",
    "-----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "marine-soldier",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cartoGRAPHs import * \n",
    "\n",
    "from func_load_data import *\n",
    "from func_visual_properties import * \n",
    "from func_calculations import * \n",
    "from func_embed_plot import * \n",
    "from func_exportVR import * "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "standing-speed",
   "metadata": {},
   "source": [
    "_____\n",
    "# 1 | HUMAN NETWORK\n",
    "_____"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "norwegian-hazard",
   "metadata": {},
   "outputs": [],
   "source": [
    "organism = 'human'\n",
    "\n",
    "G = load_graph(organism)\n",
    "\n",
    "d_centralities = load_centralities(G, organism)\n",
    "df_centralities = pd.DataFrame(d_centralities).T\n",
    "df_centralities.columns = ['degree','closeness', 'betweeness', 'eigenvector']\n",
    "essential_genes,non_ess_genes,notdefined_genes = load_essentiality(G, organism)\n",
    "\n",
    "d_gene_sym = load_genesymbols(G, organism)\n",
    "l_features = list(d_gene_sym.values())\n",
    "\n",
    "d_gene_do = pickle.load( open( \"input/d_gene_do.pkl\", \"rb\" ) )\n",
    "d_do_genes = pickle.load( open( \"input/d_do_genes.pkl\", \"rb\" ) )\n",
    "d_do_names = pickle.load( open( \"input/DO_names.pkl\", \"rb\" ) )\n",
    "d_names_do = {y:x for x,y in d_do_names.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "helpful-cargo",
   "metadata": {
    "tags": []
   },
   "source": [
    "_____ \n",
    "# 2 | FIGURE 2A + supplementary | Importance layout (with diverse node lists highlighted)\n",
    "_____"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "sensitive-hammer",
   "metadata": {},
   "outputs": [],
   "source": [
    "DM_cent = load_datamatrix(G,organism, 'importance')\n",
    "DM_cent_mod = DM_cent.round(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "female-recall",
   "metadata": {},
   "source": [
    "### 2D PORTRAIT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "suitable-silence",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_n = 20\n",
    "spr = 1.0\n",
    "md = 0.9 \n",
    "metr = 'cosine'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "greater-standing",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4min 59s, sys: 34 s, total: 5min 33s\n",
      "Wall time: 3min 23s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "\n",
    "posG_umap2D = layout_portrait_umap(G,DM_cent_mod, 2,\n",
    "                                   n_neighbors = n_n,\n",
    "                                   spread = spr,\n",
    "                                   min_dist = md,\n",
    "                                   metric = metr) \n",
    "posG = posG_umap2D"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "federal-jerusalem",
   "metadata": {
    "tags": []
   },
   "source": [
    "## VISUAL SETTINGS + PLOT PREPARATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "satisfied-martin",
   "metadata": {},
   "source": [
    "#### NODES - GENERAL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "recovered-composite",
   "metadata": {},
   "outputs": [],
   "source": [
    "opacity_nodes = 0.8\n",
    "node_edge_col = '#696969' \n",
    "node_linewidth = 1\n",
    "\n",
    "scale_factor = 0.5\n",
    "#size_nx = list(draw_node_degree(G, scale_factor).values())\n",
    "size_plotly = list(draw_node_degree(G, scale_factor/100).values())\n",
    "\n",
    "scale_factor3D = 0.025\n",
    "size3d = list(draw_node_degree_3D(G, scale_factor3D).values())\n",
    "\n",
    "nodesglow_diameter = 8.0\n",
    "nodesglow_transparency = 0.01 # 0.01"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "diagnostic-coast",
   "metadata": {},
   "source": [
    "#### ESSENTIALITY color settings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "shaped-tucson",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "color_method = 'essentiality'\n",
    "\n",
    "l_genes = essential_genes\n",
    "    \n",
    "node_col = '#00abff' \n",
    "undefined_col = '#DEDEDE '\n",
    "rest_col_nodes = '#DEDEDE '\n",
    "\n",
    "d_col_all = color_nodes_from_list(G, essential_genes, node_col)\n",
    "colors = list(d_col_all.values())\n",
    "    \n",
    "edge_color = node_col \n",
    "d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "width_edges = 0.25\n",
    "opacity_edges = 0.25"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "outdoor-lounge",
   "metadata": {},
   "source": [
    "#### DEVELOPMENTAL color settings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "stretch-documentation",
   "metadata": {},
   "outputs": [],
   "source": [
    "color_method = 'oocyte'\n",
    "\n",
    "if color_method == 'oocyte' or color_method == '2cell' or color_method == '4cell' or color_method == '8cell':\n",
    "\n",
    "    node_col = '#0181A3' \n",
    "    devgenes_selection = color_method\n",
    "    df_devgenes = pd.read_csv('input/EmExplorer/'+devgenes_selection+'_Homo-sapiens.txt', delimiter = \"\\t\")\n",
    "    \n",
    "    devgenes = list(df_devgenes['stage'])\n",
    "\n",
    "    d_devgenes_entrezid = {}\n",
    "    for i,k in d_gene_sym.items():\n",
    "        if k in devgenes:\n",
    "             d_devgenes_entrezid[k] = i \n",
    "    devgenes_entrezid = list(d_devgenes_entrezid.values())\n",
    "    \n",
    "    l_genes = devgenes_entrezid\n",
    "    \n",
    "    d_col_all = color_nodes_from_list(G, l_genes, node_col)\n",
    "    colors = list(d_col_all.values())\n",
    "\n",
    "    edge_color = node_col\n",
    "    d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "    width_edges = 0.25\n",
    "    opacity_edges = 0.5\n",
    "        \n",
    "else:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dimensional-customer",
   "metadata": {},
   "source": [
    "#### CANCER color settings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "medium-southwest",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_category = 'cancer'\n",
    "\n",
    "color_method = disease_category\n",
    "disease_genes = get_disease_genes(G, d_names_do, d_do_genes, disease_category)\n",
    "l_genes = disease_genes\n",
    "\n",
    "node_col = '#00C9FF' \n",
    "d_col_all = color_nodes_from_list(G, disease_genes, node_col)\n",
    "colors = list(d_col_all.values())\n",
    "\n",
    "edge_color = node_col \n",
    "d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "\n",
    "width_edges = 0.25\n",
    "opacity_edges = 0.25"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "arbitrary-chapel",
   "metadata": {},
   "source": [
    "#### RARE DISEASE genes color settings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "equal-appointment",
   "metadata": {},
   "outputs": [],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "tree = ET.parse('input/orphadata.xml')\n",
    "root = tree.getroot()\n",
    "\n",
    "# retrieve genes from all rare diseases from database/file  \n",
    "disorder_to_genes = {}\n",
    "for disorder in root.iter('Disorder'):\n",
    "    dis = list(disorder.attrib.values())[0]\n",
    "    for elem in disorder.findall('DisorderGeneAssociationList'):\n",
    "        genes = []\n",
    "        for disorderassoc in elem.findall('DisorderGeneAssociation'):\n",
    "            for gene in disorderassoc.findall('Gene'):\n",
    "                for reference in gene.findall('ExternalReferenceList'):\n",
    "                     for refs in reference.findall('ExternalReference'):\n",
    "                        for ensg in refs.findall('Source'):\n",
    "                            if ensg.text == 'Ensembl':\n",
    "                                for e in refs.findall('Reference'):\n",
    "                                    ensgID = (e.text)\n",
    "                            else:\n",
    "                                pass\n",
    "                            \n",
    "            genes.append(ensgID)\n",
    "    disorder_to_genes[dis] = genes\n",
    "\n",
    "    \n",
    "# get a list of all rare disease genes obtained \n",
    "all_raredisease_genes = []\n",
    "for dis,genes in disorder_to_genes.items():\n",
    "    for i in genes:\n",
    "        all_raredisease_genes.append(i)\n",
    "      \n",
    "    \n",
    "# make sure there are no duplicates\n",
    "set_all_raredisease_genes = set(all_raredisease_genes)\n",
    "\n",
    "\n",
    "# convert ensgID to entrezID and get gene symbols\n",
    "ensg_to_entrez = pd.read_csv('input/ensg_to_entrezid.txt',sep=\"\t\")\n",
    "\n",
    "ensg = list(ensg_to_entrez['From'])\n",
    "entrez = [str(int(i)) for i in ensg_to_entrez['To'][:-1]] #list(ensg_to_entrez['To'])\n",
    "d_ensg_entrez = dict(zip(ensg,entrez))\n",
    "\n",
    "\n",
    "# match ensgID with entrezID\n",
    "set_all_raredisease_genes_id = {}\n",
    "for ix,entrez in d_ensg_entrez.items():\n",
    "    if ix in set_all_raredisease_genes:\n",
    "        set_all_raredisease_genes_id[ix]=entrez\n",
    "\n",
    "        \n",
    "# color all rare disease genes  \n",
    "l_rarediseasegenes = list(set_all_raredisease_genes_id.values())\n",
    "color_method = 'rarediseasegenes'\n",
    "\n",
    "l_genes = l_rarediseasegenes\n",
    "\n",
    "node_col =  '#004EAB'#'#0A81A1' \n",
    "undefined_col = '#DEDEDE '\n",
    "rest_col_nodes = '#DEDEDE '\n",
    "\n",
    "d_col_all = color_nodes_from_list(G, l_genes, node_col)\n",
    "colors = list(d_col_all.values())\n",
    "    \n",
    "edge_color = node_col \n",
    "d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "width_edges = 0.25\n",
    "opacity_edges = 0.25"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "southern-genome",
   "metadata": {
    "tags": []
   },
   "source": [
    "## PLOT NETWORK PORTRAIT - plotly "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "ongoing-agency",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# plot nodes based on focus > separate to background/foreground \n",
    "\n",
    "posG_foreground = {}\n",
    "posG_background = {}\n",
    "for k,v in posG.items():\n",
    "    if k in l_genes:\n",
    "        posG_foreground[k]=v\n",
    "    else: \n",
    "        posG_background[k]=v\n",
    "    \n",
    "d_colors_foreground = {}\n",
    "d_colors_background = {}\n",
    "for i,c in d_col_all.items():\n",
    "    if i in posG_foreground.keys():\n",
    "        d_colors_foreground[i]=c\n",
    "    else: \n",
    "        d_colors_background[i]=c\n",
    "\n",
    "colors_foreground = list(d_colors_foreground.values())\n",
    "colors_background = list(d_colors_background.values())\n",
    "\n",
    "d_feat_foreground = {}\n",
    "d_feat_background = {}\n",
    "for k,v in d_gene_sym.items():\n",
    "    if k in posG_foreground: \n",
    "        d_feat_foreground[k]=v\n",
    "    else:\n",
    "        d_feat_background[k]=v\n",
    "        \n",
    "feat_foreground = list(d_feat_foreground.values())\n",
    "feat_background = list(d_feat_background.values())\n",
    "\n",
    "d_size_plotly = draw_node_degree(G, scale_factor/100)\n",
    "d_size_plotly_foreground = {}\n",
    "d_size_plotly_background = {}\n",
    "for k,v in d_size_plotly.items():\n",
    "    if k in posG_foreground.keys():\n",
    "        d_size_plotly_foreground[k]=v\n",
    "    else:\n",
    "        d_size_plotly_background[k]=v\n",
    "        \n",
    "size_plotly_foreground = list(d_size_plotly_foreground.values())\n",
    "size_plotly_background = list(d_size_plotly_background.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "sharing-warner",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'output_plots/2Dlayouts/2Dportrait_NetlayoutImportance_rarediseasegenes_human_onlynodes.html'"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "umap_nodes_foreground = get_trace_nodes_2D(posG_foreground, feat_foreground, colors_foreground, size_plotly_foreground, \n",
    "                                           node_linewidth)\n",
    "umap_nodes_background = get_trace_nodes_2D(posG_background, feat_background, colors_background, size_plotly_background, \n",
    "                                           node_linewidth*0.25)\n",
    "umap_nodes_background_ = get_trace_nodes_2D(posG_background, feat_background, colors_background, 0.5, node_linewidth*0.5)\n",
    "\n",
    "umap_edges = get_trace_edges_specific2D(d_edge_col, posG, linew=width_edges, opac=opacity_edges)\n",
    "data = [umap_edges, \n",
    "    umap_nodes_background_, umap_nodes_background, umap_nodes_foreground]\n",
    "\n",
    "path = 'output_plots/2Dlayouts/'\n",
    "schema = 'light' #'dark' \n",
    "fname = '2Dportrait_NetlayoutImportance_'+color_method+'_'+organism\n",
    "\n",
    "plot_2D(data,path,fname)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dominant-tomato",
   "metadata": {},
   "source": [
    "## PLOT NETWORK PORTRAIT - networkx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "requested-magnitude",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Choose edges to highlight that link inbetween 2 essential genes only \n",
    "d_edge_col = color_edges_from_nodelist_specific(G, essential_genes, node_col)\n",
    "\n",
    "# Rest edges \n",
    "d_rest_edges={}\n",
    "for e in G.edges():\n",
    "    if e not in d_edge_col.keys():\n",
    "        d_rest_edges[e] = '#f0f0f0'\n",
    "\n",
    "d_all_edges = {**d_edge_col, **d_rest_edges}\n",
    "d_all_edges_sort = {key:d_all_edges[key] for key in G.edges()}\n",
    "edge_color = list(d_all_edges_sort.values())\n",
    "\n",
    "\n",
    "\n",
    "plt.figure(figsize=(25,25))\n",
    "plt.title('2D Portrait', size=16)\n",
    "\n",
    "nx.draw_networkx_nodes(G, \n",
    "                       posG,\n",
    "                       edgecolors = 'dimgrey', \n",
    "                       linewidths = 1.2, \n",
    "                       node_color = colors, \n",
    "                       node_size = size_nx, \n",
    "                       alpha = 0.8)\n",
    "\n",
    "nx.draw_networkx_edges(G, posG, width = 0.1, \n",
    "                       edge_color = edge_color, \n",
    "                      alpha = 0.25)\n",
    "\n",
    "plt.box(False)\n",
    "\n",
    "fname = 'output_plots/2Dlayouts/2Dportrait_Fig2a_'+organism\n",
    "\n",
    "# PNG\n",
    "#plt.savefig(fname+'.png',dpi=300)\n",
    "\n",
    "# PDF \n",
    "plt.savefig(fname+'.pdf',format='pdf')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "focal-wagon",
   "metadata": {
    "tags": []
   },
   "source": [
    "_____ \n",
    "# 2 | FIGURE 2B - disease layout\n",
    "_____"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "better-passage",
   "metadata": {},
   "outputs": [],
   "source": [
    "DM_disease = load_datamatrix(G,organism, 'funct-dis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "certain-calvin",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>389289</th>\n",
       "      <th>4524</th>\n",
       "      <th>4353</th>\n",
       "      <th>27127</th>\n",
       "      <th>5925</th>\n",
       "      <th>10735</th>\n",
       "      <th>26259</th>\n",
       "      <th>1029</th>\n",
       "      <th>9649</th>\n",
       "      <th>1630</th>\n",
       "      <th>...</th>\n",
       "      <th>65125</th>\n",
       "      <th>6787</th>\n",
       "      <th>90678</th>\n",
       "      <th>10661</th>\n",
       "      <th>56479</th>\n",
       "      <th>63910</th>\n",
       "      <th>7287</th>\n",
       "      <th>11173</th>\n",
       "      <th>9962</th>\n",
       "      <th>5973</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>389289</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.835970</td>\n",
       "      <td>0.747354</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.780911</td>\n",
       "      <td>0.438049</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.807550</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.634852</td>\n",
       "      <td>...</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.817426</td>\n",
       "      <td>0.951205</td>\n",
       "      <td>0.894591</td>\n",
       "      <td>0.855662</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.951887</td>\n",
       "      <td>0.855662</td>\n",
       "      <td>0.845697</td>\n",
       "      <td>0.938454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4524</th>\n",
       "      <td>0.835970</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.502703</td>\n",
       "      <td>0.835970</td>\n",
       "      <td>0.706513</td>\n",
       "      <td>0.800283</td>\n",
       "      <td>0.835970</td>\n",
       "      <td>0.742198</td>\n",
       "      <td>0.835970</td>\n",
       "      <td>0.810332</td>\n",
       "      <td>...</td>\n",
       "      <td>0.808632</td>\n",
       "      <td>0.850262</td>\n",
       "      <td>0.967985</td>\n",
       "      <td>0.896258</td>\n",
       "      <td>0.857946</td>\n",
       "      <td>0.863308</td>\n",
       "      <td>0.834270</td>\n",
       "      <td>0.834270</td>\n",
       "      <td>0.898759</td>\n",
       "      <td>0.959619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4353</th>\n",
       "      <td>0.747354</td>\n",
       "      <td>0.502703</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.747354</td>\n",
       "      <td>0.750917</td>\n",
       "      <td>0.858025</td>\n",
       "      <td>0.747354</td>\n",
       "      <td>0.789306</td>\n",
       "      <td>0.747354</td>\n",
       "      <td>0.830869</td>\n",
       "      <td>...</td>\n",
       "      <td>0.747354</td>\n",
       "      <td>0.815494</td>\n",
       "      <td>0.963016</td>\n",
       "      <td>0.920106</td>\n",
       "      <td>0.890601</td>\n",
       "      <td>0.789462</td>\n",
       "      <td>0.890601</td>\n",
       "      <td>0.817669</td>\n",
       "      <td>0.844064</td>\n",
       "      <td>0.937803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27127</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.835970</td>\n",
       "      <td>0.747354</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.780911</td>\n",
       "      <td>0.438049</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.807550</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.634852</td>\n",
       "      <td>...</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.817426</td>\n",
       "      <td>0.951205</td>\n",
       "      <td>0.894591</td>\n",
       "      <td>0.855662</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.951887</td>\n",
       "      <td>0.855662</td>\n",
       "      <td>0.845697</td>\n",
       "      <td>0.938454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5925</th>\n",
       "      <td>0.780911</td>\n",
       "      <td>0.706513</td>\n",
       "      <td>0.750917</td>\n",
       "      <td>0.780911</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.712726</td>\n",
       "      <td>0.780911</td>\n",
       "      <td>0.578363</td>\n",
       "      <td>0.780911</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.908713</td>\n",
       "      <td>0.920000</td>\n",
       "      <td>0.967929</td>\n",
       "      <td>0.930718</td>\n",
       "      <td>0.810263</td>\n",
       "      <td>0.817426</td>\n",
       "      <td>0.862968</td>\n",
       "      <td>0.873509</td>\n",
       "      <td>0.932388</td>\n",
       "      <td>0.959548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63910</th>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.863308</td>\n",
       "      <td>0.789462</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.817426</td>\n",
       "      <td>0.906341</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.871700</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.939142</td>\n",
       "      <td>...</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.817426</td>\n",
       "      <td>0.902410</td>\n",
       "      <td>0.789181</td>\n",
       "      <td>0.711325</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.903775</td>\n",
       "      <td>0.711325</td>\n",
       "      <td>0.691393</td>\n",
       "      <td>0.876909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7287</th>\n",
       "      <td>0.951887</td>\n",
       "      <td>0.834270</td>\n",
       "      <td>0.890601</td>\n",
       "      <td>0.951887</td>\n",
       "      <td>0.862968</td>\n",
       "      <td>0.972963</td>\n",
       "      <td>0.951887</td>\n",
       "      <td>0.907407</td>\n",
       "      <td>0.951887</td>\n",
       "      <td>0.964864</td>\n",
       "      <td>...</td>\n",
       "      <td>0.879719</td>\n",
       "      <td>0.894591</td>\n",
       "      <td>0.929570</td>\n",
       "      <td>0.939142</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.903775</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.910913</td>\n",
       "      <td>0.964467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11173</th>\n",
       "      <td>0.855662</td>\n",
       "      <td>0.834270</td>\n",
       "      <td>0.817669</td>\n",
       "      <td>0.855662</td>\n",
       "      <td>0.873509</td>\n",
       "      <td>0.918889</td>\n",
       "      <td>0.855662</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>0.855662</td>\n",
       "      <td>0.947295</td>\n",
       "      <td>...</td>\n",
       "      <td>0.639156</td>\n",
       "      <td>0.841886</td>\n",
       "      <td>0.915485</td>\n",
       "      <td>0.817426</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.711325</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.732739</td>\n",
       "      <td>0.840099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9962</th>\n",
       "      <td>0.845697</td>\n",
       "      <td>0.898759</td>\n",
       "      <td>0.844064</td>\n",
       "      <td>0.845697</td>\n",
       "      <td>0.932388</td>\n",
       "      <td>0.913289</td>\n",
       "      <td>0.845697</td>\n",
       "      <td>0.910913</td>\n",
       "      <td>0.845697</td>\n",
       "      <td>0.943656</td>\n",
       "      <td>...</td>\n",
       "      <td>0.845697</td>\n",
       "      <td>0.830969</td>\n",
       "      <td>0.909649</td>\n",
       "      <td>0.804820</td>\n",
       "      <td>0.732739</td>\n",
       "      <td>0.691393</td>\n",
       "      <td>0.910913</td>\n",
       "      <td>0.732739</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.886039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5973</th>\n",
       "      <td>0.938454</td>\n",
       "      <td>0.959619</td>\n",
       "      <td>0.937803</td>\n",
       "      <td>0.938454</td>\n",
       "      <td>0.959548</td>\n",
       "      <td>0.965414</td>\n",
       "      <td>0.938454</td>\n",
       "      <td>0.976311</td>\n",
       "      <td>0.938454</td>\n",
       "      <td>0.977527</td>\n",
       "      <td>...</td>\n",
       "      <td>0.907681</td>\n",
       "      <td>0.932580</td>\n",
       "      <td>0.963963</td>\n",
       "      <td>0.922150</td>\n",
       "      <td>0.893400</td>\n",
       "      <td>0.876909</td>\n",
       "      <td>0.964467</td>\n",
       "      <td>0.840099</td>\n",
       "      <td>0.886039</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6538 rows × 6538 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          389289      4524      4353     27127      5925     10735     26259  \\\n",
       "389289  0.000000  0.835970  0.747354  0.000000  0.780911  0.438049  0.000000   \n",
       "4524    0.835970  0.000000  0.502703  0.835970  0.706513  0.800283  0.835970   \n",
       "4353    0.747354  0.502703  0.000000  0.747354  0.750917  0.858025  0.747354   \n",
       "27127   0.000000  0.835970  0.747354  0.000000  0.780911  0.438049  0.000000   \n",
       "5925    0.780911  0.706513  0.750917  0.780911  0.000000  0.712726  0.780911   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "63910   0.833333  0.863308  0.789462  0.833333  0.817426  0.906341  0.833333   \n",
       "7287    0.951887  0.834270  0.890601  0.951887  0.862968  0.972963  0.951887   \n",
       "11173   0.855662  0.834270  0.817669  0.855662  0.873509  0.918889  0.855662   \n",
       "9962    0.845697  0.898759  0.844064  0.845697  0.932388  0.913289  0.845697   \n",
       "5973    0.938454  0.959619  0.937803  0.938454  0.959548  0.965414  0.938454   \n",
       "\n",
       "            1029      9649      1630  ...     65125      6787     90678  \\\n",
       "389289  0.807550  0.000000  0.634852  ...  0.916667  0.817426  0.951205   \n",
       "4524    0.742198  0.835970  0.810332  ...  0.808632  0.850262  0.967985   \n",
       "4353    0.789306  0.747354  0.830869  ...  0.747354  0.815494  0.963016   \n",
       "27127   0.807550  0.000000  0.634852  ...  0.916667  0.817426  0.951205   \n",
       "5925    0.578363  0.780911  0.800000  ...  0.908713  0.920000  0.967929   \n",
       "...          ...       ...       ...  ...       ...       ...       ...   \n",
       "63910   0.871700  0.833333  0.939142  ...  0.833333  0.817426  0.902410   \n",
       "7287    0.907407  0.951887  0.964864  ...  0.879719  0.894591  0.929570   \n",
       "11173   0.944444  0.855662  0.947295  ...  0.639156  0.841886  0.915485   \n",
       "9962    0.910913  0.845697  0.943656  ...  0.845697  0.830969  0.909649   \n",
       "5973    0.976311  0.938454  0.977527  ...  0.907681  0.932580  0.963963   \n",
       "\n",
       "           10661     56479     63910      7287     11173      9962      5973  \n",
       "389289  0.894591  0.855662  0.833333  0.951887  0.855662  0.845697  0.938454  \n",
       "4524    0.896258  0.857946  0.863308  0.834270  0.834270  0.898759  0.959619  \n",
       "4353    0.920106  0.890601  0.789462  0.890601  0.817669  0.844064  0.937803  \n",
       "27127   0.894591  0.855662  0.833333  0.951887  0.855662  0.845697  0.938454  \n",
       "5925    0.930718  0.810263  0.817426  0.862968  0.873509  0.932388  0.959548  \n",
       "...          ...       ...       ...       ...       ...       ...       ...  \n",
       "63910   0.789181  0.711325  0.000000  0.903775  0.711325  0.691393  0.876909  \n",
       "7287    0.939142  0.666667  0.903775  0.000000  0.875000  0.910913  0.964467  \n",
       "11173   0.817426  0.750000  0.711325  0.875000  0.000000  0.732739  0.840099  \n",
       "9962    0.804820  0.732739  0.691393  0.910913  0.732739  0.000000  0.886039  \n",
       "5973    0.922150  0.893400  0.876909  0.964467  0.840099  0.886039  0.000000  \n",
       "\n",
       "[6538 rows x 6538 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DM_disease"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "happy-monitor",
   "metadata": {
    "tags": []
   },
   "source": [
    "## DISEASE GENES - Neurofibromatosis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "subtle-parking",
   "metadata": {},
   "source": [
    "#### 1. | NF1+neighbor genes / color settings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "sound-toddler",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_nf = 'input/neurofibromatosis/'\n",
    "G_neuro = nx.read_edgelist(path_nf+'nf_edges.txt', delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "designed-electronics",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene of interest\n",
    "nf_goi_ = str(int(np.loadtxt(path_nf+'nf_gene_of_interest.txt', unpack=False)))\n",
    "nf_goi = list(nf_goi_.split(\" \"))\n",
    "\n",
    "# pathway genes\n",
    "nf_path_ = np.loadtxt(path_nf+'nf_pathway_genes.txt', unpack=False)\n",
    "nf_pathw = []\n",
    "for i in nf_path_:\n",
    "    nf_pathw.append(str(int(i)))\n",
    "    \n",
    "# NF1 neighbors \n",
    "df_nf_neigh = pd.read_csv(path_nf+'nf_neighbors.txt', header = None,index_col=False)\n",
    "nf_neigh = [str(i) for i in list(df_nf_neigh[0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "cleared-backup",
   "metadata": {},
   "outputs": [],
   "source": [
    "enhance_genes = nf_goi + nf_pathw + nf_neigh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mexican-magic",
   "metadata": {},
   "source": [
    "# How to enhance disease associated genes of NF in disease matrix "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "sealed-billy",
   "metadata": {},
   "outputs": [],
   "source": [
    "DM_global = load_datamatrix(G,organism, 'global')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "determined-sheep",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>66008</th>\n",
       "      <th>8473</th>\n",
       "      <th>2561</th>\n",
       "      <th>3759</th>\n",
       "      <th>22906</th>\n",
       "      <th>4928</th>\n",
       "      <th>1994</th>\n",
       "      <th>8481</th>\n",
       "      <th>81610</th>\n",
       "      <th>51361</th>\n",
       "      <th>...</th>\n",
       "      <th>10838</th>\n",
       "      <th>8001</th>\n",
       "      <th>51351</th>\n",
       "      <th>1551</th>\n",
       "      <th>51458</th>\n",
       "      <th>143903</th>\n",
       "      <th>10861</th>\n",
       "      <th>51471</th>\n",
       "      <th>221044</th>\n",
       "      <th>29965</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66008</th>\n",
       "      <td>0.900242</td>\n",
       "      <td>0.000687</td>\n",
       "      <td>0.012865</td>\n",
       "      <td>0.006433</td>\n",
       "      <td>0.005347</td>\n",
       "      <td>0.000460</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>0.000345</td>\n",
       "      <td>0.003464</td>\n",
       "      <td>0.002905</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8473</th>\n",
       "      <td>0.002519</td>\n",
       "      <td>0.900245</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.005316</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2561</th>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.900367</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3759</th>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.900389</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22906</th>\n",
       "      <td>0.002525</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.900120</td>\n",
       "      <td>0.000459</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143903</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.900070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10861</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.900070</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51471</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.900052</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221044</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.90003</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29965</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.900063</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16376 rows × 16376 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           66008      8473      2561      3759     22906      4928      1994  \\\n",
       "66008   0.900242  0.000687  0.012865  0.006433  0.005347  0.000460  0.000051   \n",
       "8473    0.002519  0.900245  0.000036  0.000018  0.005316  0.000005  0.000002   \n",
       "2561    0.002502  0.000002  0.900367  0.000018  0.000015  0.000002  0.000000   \n",
       "3759    0.002502  0.000002  0.000036  0.900389  0.000015  0.000002  0.000000   \n",
       "22906   0.002525  0.000685  0.000036  0.000018  0.900120  0.000459  0.000000   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "143903  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "10861   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "51471   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "221044  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "29965   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "            8481     81610     51361  ...  10838      8001     51351  1551  \\\n",
       "66008   0.000345  0.003464  0.002905  ...    0.0  0.000000  0.000000   0.0   \n",
       "8473    0.000003  0.000012  0.000008  ...    0.0  0.000000  0.000000   0.0   \n",
       "2561    0.000001  0.000010  0.000008  ...    0.0  0.000000  0.000001   0.0   \n",
       "3759    0.000001  0.000010  0.000008  ...    0.0  0.000001  0.000000   0.0   \n",
       "22906   0.000001  0.000010  0.000009  ...    0.0  0.000000  0.000000   0.0   \n",
       "...          ...       ...       ...  ...    ...       ...       ...   ...   \n",
       "143903  0.000000  0.000000  0.000000  ...    0.0  0.000000  0.000070   0.0   \n",
       "10861   0.000000  0.000000  0.000000  ...    0.0  0.000000  0.000070   0.0   \n",
       "51471   0.000000  0.000000  0.000000  ...    0.0  0.000000  0.000000   0.0   \n",
       "221044  0.000000  0.000000  0.000000  ...    0.0  0.000000  0.000000   0.0   \n",
       "29965   0.000000  0.000000  0.000000  ...    0.0  0.000000  0.000000   0.0   \n",
       "\n",
       "        51458    143903     10861     51471   221044     29965  \n",
       "66008     0.0  0.000000  0.000000  0.000000  0.00000  0.000000  \n",
       "8473      0.0  0.000000  0.000000  0.000000  0.00000  0.000000  \n",
       "2561      0.0  0.000001  0.000001  0.000000  0.00000  0.000000  \n",
       "3759      0.0  0.000000  0.000000  0.000000  0.00000  0.000000  \n",
       "22906     0.0  0.000000  0.000000  0.000000  0.00000  0.000000  \n",
       "...       ...       ...       ...       ...      ...       ...  \n",
       "143903    0.0  0.900070  0.000070  0.000000  0.00000  0.000000  \n",
       "10861     0.0  0.000070  0.900070  0.000000  0.00000  0.000000  \n",
       "51471     0.0  0.000000  0.000000  0.900052  0.00000  0.000000  \n",
       "221044    0.0  0.000000  0.000000  0.000000  0.90003  0.000000  \n",
       "29965     0.0  0.000000  0.000000  0.000000  0.00000  0.900063  \n",
       "\n",
       "[16376 rows x 16376 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DM_global"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sharing-differential",
   "metadata": {},
   "source": [
    "### Estimate Turning point of Enhancing Factor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "productive-organizer",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1 | get all values from merged (struc + funct) dataframe \n",
    "\n",
    "df_merge_list = DF_merge.values.tolist()\n",
    "l_merge_all = []\n",
    "for i in df_merge_list:\n",
    "    l_merge_all.extend(i)\n",
    "    \n",
    "# 2 | get a set of values for x-axis \n",
    "\n",
    "set_all_values = set(l_merge_all)\n",
    "\n",
    "# sort from low to high \n",
    "set_all_sorted = sorted(set_all_values)\n",
    "\n",
    "\n",
    "# 3 | get the count of each value for y-axis \n",
    "from collections import Counter\n",
    "\n",
    "occurrence_counted = Counter(l_merge_all)\n",
    "\n",
    "occurrence_counted_sorted = {}\n",
    "for i in set_all_sorted:\n",
    "    for k,v in occurrence_counted.items():\n",
    "         if i == k:\n",
    "            occurrence_counted_sorted[i]=v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "indian-overall",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "instant-programmer",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "comprehensive-spank",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>66008</th>\n",
       "      <th>8473</th>\n",
       "      <th>2561</th>\n",
       "      <th>3759</th>\n",
       "      <th>22906</th>\n",
       "      <th>4928</th>\n",
       "      <th>1994</th>\n",
       "      <th>8481</th>\n",
       "      <th>81610</th>\n",
       "      <th>51361</th>\n",
       "      <th>...</th>\n",
       "      <th>8001</th>\n",
       "      <th>51351</th>\n",
       "      <th>1551</th>\n",
       "      <th>51458</th>\n",
       "      <th>143903</th>\n",
       "      <th>10861</th>\n",
       "      <th>51471</th>\n",
       "      <th>221044</th>\n",
       "      <th>29965</th>\n",
       "      <th>NF1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66008</th>\n",
       "      <td>0.900242</td>\n",
       "      <td>0.002519</td>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.002525</td>\n",
       "      <td>0.002517</td>\n",
       "      <td>0.002512</td>\n",
       "      <td>0.002511</td>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.002501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8473</th>\n",
       "      <td>0.000687</td>\n",
       "      <td>0.900245</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2561</th>\n",
       "      <td>0.012865</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.900367</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3759</th>\n",
       "      <td>0.006433</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.900389</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22906</th>\n",
       "      <td>0.005347</td>\n",
       "      <td>0.005316</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.900120</td>\n",
       "      <td>0.005314</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143903</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00007</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.90007</td>\n",
       "      <td>0.00007</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10861</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00007</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00007</td>\n",
       "      <td>0.90007</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51471</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.900052</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221044</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.90003</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29965</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.900063</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16376 rows × 16377 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           66008      8473      2561      3759     22906      4928      1994  \\\n",
       "66008   0.900242  0.002519  0.002502  0.002502  0.002525  0.002517  0.002512   \n",
       "8473    0.000687  0.900245  0.000002  0.000002  0.000685  0.000008  0.000031   \n",
       "2561    0.012865  0.000036  0.900367  0.000036  0.000036  0.000065  0.000037   \n",
       "3759    0.006433  0.000018  0.000018  0.900389  0.000018  0.000027  0.000041   \n",
       "22906   0.005347  0.005316  0.000015  0.000015  0.900120  0.005314  0.000026   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "143903  0.000000  0.000000  0.000001  0.000000  0.000000  0.000000  0.000004   \n",
       "10861   0.000000  0.000000  0.000001  0.000000  0.000000  0.000000  0.000004   \n",
       "51471   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000054   \n",
       "221044  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000003   \n",
       "29965   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000002   \n",
       "\n",
       "            8481     81610     51361  ...  8001    51351  1551  51458  \\\n",
       "66008   0.002511  0.002502  0.002501  ...   0.0  0.00000   0.0    0.0   \n",
       "8473    0.000006  0.000002  0.000002  ...   0.0  0.00000   0.0    0.0   \n",
       "2561    0.000036  0.000036  0.000036  ...   0.0  0.00000   0.0    0.0   \n",
       "3759    0.000018  0.000018  0.000018  ...   0.0  0.00000   0.0    0.0   \n",
       "22906   0.000015  0.000015  0.000016  ...   0.0  0.00000   0.0    0.0   \n",
       "...          ...       ...       ...  ...   ...      ...   ...    ...   \n",
       "143903  0.000000  0.000000  0.000000  ...   0.0  0.00007   0.0    0.0   \n",
       "10861   0.000000  0.000000  0.000000  ...   0.0  0.00007   0.0    0.0   \n",
       "51471   0.000000  0.000000  0.000000  ...   0.0  0.00000   0.0    0.0   \n",
       "221044  0.000000  0.000000  0.000000  ...   0.0  0.00000   0.0    0.0   \n",
       "29965   0.000046  0.000000  0.000000  ...   0.0  0.00000   0.0    0.0   \n",
       "\n",
       "         143903    10861     51471   221044     29965  NF1  \n",
       "66008   0.00000  0.00000  0.000000  0.00000  0.000000  0.0  \n",
       "8473    0.00000  0.00000  0.000000  0.00000  0.000000  0.0  \n",
       "2561    0.00000  0.00000  0.000000  0.00000  0.000000  0.0  \n",
       "3759    0.00000  0.00000  0.000000  0.00000  0.000000  0.0  \n",
       "22906   0.00000  0.00000  0.000000  0.00000  0.000000  0.0  \n",
       "...         ...      ...       ...      ...       ...  ...  \n",
       "143903  0.90007  0.00007  0.000000  0.00000  0.000000  0.0  \n",
       "10861   0.00007  0.90007  0.000000  0.00000  0.000000  0.0  \n",
       "51471   0.00000  0.00000  0.900052  0.00000  0.000000  0.0  \n",
       "221044  0.00000  0.00000  0.000000  0.90003  0.000000  0.0  \n",
       "29965   0.00000  0.00000  0.000000  0.00000  0.900063  0.0  \n",
       "\n",
       "[16376 rows x 16377 columns]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d_vec_disease = {}\n",
    "for i in G.nodes():\n",
    "    if i in enhance_genes:\n",
    "        d_vec_disease[i]=enhance_factor\n",
    "    else:\n",
    "        d_vec_disease[i]=0\n",
    "        \n",
    "d_vec_disease_sorted = {key:d_vec_disease[key] for key in G.nodes()}\n",
    "\n",
    "DF_functional_scal = pd.DataFrame(list(d_vec_disease_sorted.values()))\n",
    "DF_functional_scal.index = list(G.nodes())\n",
    "DF_functional_scal.columns = ['NF1']\n",
    "\n",
    "DF_functional_scal = DF_functional_scal.T\n",
    "DF_structural = DM_global \n",
    "\n",
    "DF_merge = pd.concat([DF_structural, \n",
    "                      DF_functional_scal]).T\n",
    "\n",
    "DF_merge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "joint-holder",
   "metadata": {},
   "source": [
    "### 2D PORTRAIT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "scheduled-surveillance",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_n = 20\n",
    "spr = 1\n",
    "md = 0.9 \n",
    "metr = 'cosine'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "occupational-interest",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min 13s, sys: 7.68 s, total: 3min 21s\n",
      "Wall time: 1min 55s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "\n",
    "posG_umap2D = layout_portrait_umap(G,DF_merge, 2,\n",
    "                                   n_neighbors = n_n,\n",
    "                                   spread = spr,\n",
    "                                   min_dist = md,\n",
    "                                   metric = metr) \n",
    "posG = posG_umap2D"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abroad-chorus",
   "metadata": {
    "tags": []
   },
   "source": [
    "## VISUAL SETTINGS + PLOT PREPARATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "severe-present",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. disease layout \n",
    "# highlight NF1 + edges to neighbors + neighbor genes \n",
    "# other genes are grey; no other edges "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "solid-secretary",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. highlighting same genes \n",
    "# add/show all edges inbetween disease genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "developmental-patch",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3. show NF1 links to other disease clusters (4 of eva's example)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "impaired-borough",
   "metadata": {},
   "source": [
    "#### NODES - GENERAL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "stretch-upgrade",
   "metadata": {},
   "outputs": [],
   "source": [
    "opacity_nodes = 0.8\n",
    "node_edge_col = '#696969' \n",
    "node_linewidth = 1\n",
    "\n",
    "scale_factor = 0.5\n",
    "#size_nx = list(draw_node_degree(G, scale_factor).values())\n",
    "size_plotly = list(draw_node_degree(G, scale_factor/100).values())\n",
    "\n",
    "scale_factor3D = 0.025\n",
    "size3d = list(draw_node_degree_3D(G, scale_factor3D).values())\n",
    "\n",
    "nodesglow_diameter = 8.0\n",
    "nodesglow_transparency = 0.01 # 0.01"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "impaired-setting",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "color_method = 'NF1'\n",
    "l_genes = enhance_genes\n",
    "\n",
    "node_col = '#FF8E00' \n",
    "undefined_col = '#DEDEDE'\n",
    "rest_col_nodes = '#DEDEDE'\n",
    "\n",
    "d_col_all = color_nodes_from_list(G, l_genes, node_col)\n",
    "colors = list(d_col_all.values())\n",
    "    \n",
    "edge_color = node_col \n",
    "d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "width_edges = 0.25\n",
    "opacity_edges = 0.5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "informative-immunology",
   "metadata": {
    "tags": []
   },
   "source": [
    "## PLOT NETWORK PORTRAIT - plotly "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "white-lodging",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# plot nodes based on focus > separate to background/foreground \n",
    "\n",
    "posG_foreground = {}\n",
    "posG_background = {}\n",
    "for k,v in posG.items():\n",
    "    if k in l_genes:\n",
    "        posG_foreground[k]=v\n",
    "    else: \n",
    "        posG_background[k]=v\n",
    "    \n",
    "d_colors_foreground = {}\n",
    "d_colors_background = {}\n",
    "for i,c in d_col_all.items():\n",
    "    if i in posG_foreground.keys():\n",
    "        d_colors_foreground[i]=c\n",
    "    else: \n",
    "        d_colors_background[i]=c\n",
    "\n",
    "colors_foreground = list(d_colors_foreground.values())\n",
    "colors_background = list(d_colors_background.values())\n",
    "\n",
    "d_feat_foreground = {}\n",
    "d_feat_background = {}\n",
    "for k,v in d_gene_sym.items():\n",
    "    if k in posG_foreground: \n",
    "        d_feat_foreground[k]=v\n",
    "    else:\n",
    "        d_feat_background[k]=v\n",
    "        \n",
    "feat_foreground = list(d_feat_foreground.values())\n",
    "feat_background = list(d_feat_background.values())\n",
    "\n",
    "d_size_plotly = draw_node_degree(G, scale_factor/100)\n",
    "d_size_plotly_foreground = {}\n",
    "d_size_plotly_background = {}\n",
    "for k,v in d_size_plotly.items():\n",
    "    if k in posG_foreground.keys():\n",
    "        d_size_plotly_foreground[k]=v\n",
    "    else:\n",
    "        d_size_plotly_background[k]=v\n",
    "        \n",
    "size_plotly_foreground = list(d_size_plotly_foreground.values())\n",
    "size_plotly_background = list(d_size_plotly_background.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "olympic-aerospace",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'output_plots/2Dlayouts/2Dportrait_NetlayoutDisease_NF1_human.html'"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "umap_nodes_foreground = get_trace_nodes_2D(posG_foreground, feat_foreground, colors_foreground, size_plotly_foreground, node_linewidth)\n",
    "umap_nodes_background = get_trace_nodes_2D(posG_background, feat_background, colors_background, size_plotly_background, node_linewidth*0.25)\n",
    "umap_nodes_background_ = get_trace_nodes_2D(posG_background, feat_background, colors_background, 0.5, node_linewidth*0.5)\n",
    "\n",
    "umap_edges = get_trace_edges_specific2D(d_edge_col, posG, linew=width_edges, opac=opacity_edges)\n",
    "data = [umap_edges, umap_nodes_background_, umap_nodes_background, umap_nodes_foreground]\n",
    "\n",
    "path = 'output_plots/2Dlayouts/'\n",
    "schema = 'light' #'dark' \n",
    "fname = '2Dportrait_NetlayoutDisease_'+color_method+'_'+organism\n",
    "\n",
    "plot_2D(data,path,fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "noticed-justice",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "selective-designer",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "demonstrated-ensemble",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dutch-criterion",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "welcome-domain",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "municipal-lesbian",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fifteen-hunter",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "collaborative-berkeley",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "systematic-prize",
   "metadata": {
    "tags": []
   },
   "source": [
    "### TOPOGRAPHIC MAP "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "fluid-iraqi",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------------------------------------------\n",
    "#\n",
    "# Choose a z-parameter e.g. essentiality, centravg, disease \n",
    "# or choose any dictionary with nodeID: z-value\n",
    "# by setting d_z = {nodeID: val, ...}\n",
    "#\n",
    "# ---------------------------------------------------------------------------\n",
    "z_feat = 'disease'\n",
    "# ---------------------------------------------------------------------------\n",
    "\n",
    "if z_feat == 'essentiality':\n",
    "\n",
    "    ''' Essentiality state of each node displayed on layers of \"essential\", \"non-essential\", \"non-categorized\" ''' \n",
    "\n",
    "    value_ess = 5\n",
    "    value_noness = 3 \n",
    "    value_notdef = 1\n",
    "\n",
    "    d_ess = {}\n",
    "    for i in essential_genes:\n",
    "        d_ess[i] = value_ess\n",
    "    d_noness = {}\n",
    "    for i in non_ess_genes:\n",
    "        d_noness[i] = value_noness\n",
    "    d_notdef = {}\n",
    "    for i in notdefined_genes:\n",
    "        d_notdef[i] = value_notdef\n",
    "\n",
    "    d_alless_unsort = {**d_ess,**d_noness,**d_notdef}\n",
    "    d_z = {key:d_alless_unsort[key] for key in G.nodes()}\n",
    "    \n",
    "    l_genes = essential_genes\n",
    "    \n",
    "    node_col = '#00abff' \n",
    "    undefined_col = '#d3d3d3'\n",
    "    rest_col_nodes = '#d3d3d3'\n",
    "\n",
    "    d_col_all = color_nodes_from_list(G, essential_genes, node_col)\n",
    "    colors = list(d_col_all.values())\n",
    "    \n",
    "    edge_color = '#ACACAC' \n",
    "    d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "    width_edges = 0.15\n",
    "    opacity_edges = 0.1\n",
    "    \n",
    "    umap_nodes = get_trace_nodes_3D(posG, l_features, colors, size, opacity_nodes)\n",
    "    umap_edges = get_trace_edges_specific3D(d_edge_col, posG, linew=width_edges, opac=opacity_edges)  \n",
    "    data = [umap_edges, umap_nodes]\n",
    "    \n",
    "    \n",
    "elif z_feat == 'centravg':\n",
    "    \n",
    "    ''' Average of Centrality Metrics displayed in z-height ''' \n",
    "    \n",
    "    d_clos = {}\n",
    "    for k,v in d_centralities.items():\n",
    "        d_clos[k] = v[1]\n",
    "\n",
    "    d_nodecolors = d_clos \n",
    "    col_pal = 'YlOrRd'\n",
    "\n",
    "    d_colors = color_nodes_from_dict(G, d_nodecolors, palette = col_pal)\n",
    "    colors = list(d_colors.values())\n",
    "    \n",
    "    edge_color = '#ACACAC' \n",
    "    width_edges = 0.15\n",
    "    opacity_edges = 0.1\n",
    "    \n",
    "    umap_nodes = get_trace_nodes_3D(posG, l_features, colors, size, opacity_nodes)\n",
    "    umap_edges = get_trace_edges_3D(G, posG, edge_color, opac = 0.05, linewidth = 0.5)\n",
    "    data = [umap_edges, umap_nodes]\n",
    "    \n",
    "    \n",
    "elif z_feat == 'disease':\n",
    "    \n",
    "    ''' Disease count associated to a node displayed in z-height ''' \n",
    "\n",
    "    DM_feature  = pd.read_csv('input/Features_Disease_Dataframe_Human.csv', index_col=0)\n",
    "\n",
    "    # set gene list (= G.nodes())\n",
    "    genes_assoc = []\n",
    "    for i in DM_feature.index:\n",
    "        genes_assoc.append(str(i))\n",
    "\n",
    "    genes_notassoc = [] \n",
    "    for g in G.nodes():\n",
    "        if g not in genes_assoc:\n",
    "            genes_notassoc.append(g)\n",
    "\n",
    "    features_counted = (DM_feature == 1).astype(int).sum(axis=1)\n",
    "    d_features_counted = dict(zip(genes_assoc, list(features_counted)))\n",
    "    \n",
    "    d_rest = {}\n",
    "    for i in genes_notassoc: \n",
    "        d_rest[i] = -20\n",
    "\n",
    "    d_param_unsorted = {**d_features_counted, **d_rest}\n",
    "    d_z = {key:d_param_unsorted[key] for key in G.nodes()}\n",
    "    \n",
    "    d_clos = {}\n",
    "    for k,v in d_centralities.items():\n",
    "        d_clos[k] = v[1]\n",
    "\n",
    "    d_nodecolors = d_clos \n",
    "    col_pal = 'YlOrRd'\n",
    "\n",
    "    d_colors = color_nodes_from_dict(G, d_nodecolors, palette = col_pal)\n",
    "    colors = list(d_colors.values())\n",
    "    edge_color = '#ACACAC'\n",
    "    width_edges = 0.15\n",
    "    opacity_edges = 0.1\n",
    "\n",
    "    umap_nodes = get_trace_nodes_3D(posG, l_features, colors, size, opacity_nodes)\n",
    "    umap_edges = get_trace_edges_3D(G, posG, edge_color, opac = 0.05, linewidth = 0.5)\n",
    "    data = [umap_edges, umap_nodes]\n",
    "    \n",
    "else:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nasty-marina",
   "metadata": {},
   "source": [
    "#### PLOT "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "assisted-message",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'output_plots/Topographic/topographic_z_disease_local_human_light.html'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ---------------------------------------------------------------------------\n",
    "#\n",
    "# Choose a 2D layout as basis\n",
    "#\n",
    "# 1. either calculate one e.g.: \n",
    "# posG_2Dportrait = layout_portrait_umap(G,DM,2,n_neighbors=20, spread=1, min_dist=0.0, metric='cosine') \n",
    "# ---------------------------------------------------------------------------\n",
    "#\n",
    "# 2. or use a precalculated 2D layout \n",
    "posG_2D = posG_umap2D\n",
    "# ---------------------------------------------------------------------------\n",
    "\n",
    "posG = layout_topographic(posG_2D, d_z)\n",
    "\n",
    "path = 'output_plots/Topographic/'\n",
    "schema = 'light' #'dark'\n",
    "fname = 'topographic_z_'+z_feat+'_'+netlayout+'_'+organism+'_'+schema\n",
    "\n",
    "plot_3D(data,path,fname, schema)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "finite-directive",
   "metadata": {},
   "source": [
    "### GEODESIC MAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "powered-chassis",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Spherical Shells: 4\n"
     ]
    }
   ],
   "source": [
    "# ---------------------------------------------------------------------------\n",
    "#\n",
    "# Choose a radius-parameter e.g. autocore, essentiality\n",
    "# or choose any dictionary with nodeID: z-value\n",
    "# by setting d_radius = {nodeID: val, ...}\n",
    "#\n",
    "# ---------------------------------------------------------------------------\n",
    "r_feat = 'autocore'\n",
    "# ---------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "if r_feat == 'autocore':\n",
    "\n",
    "    # Load Gene Information for Gene lists \n",
    "    variants = pd.read_csv('input/julias_variants_mod.txt')\n",
    "    variant_list = [str(i) for i in list(variants['variants'])]\n",
    "    variant_sym = pd.read_csv('input/julias_variants_sym.txt')\n",
    "    genes_sym_list = [str(i) for i in list(variant_sym['variant_sym'])]\n",
    "    d_gene_sym_variants = dict(zip(variant_list,genes_sym_list))\n",
    "\n",
    "    df_seeds = pd.read_csv('input/seeds_from_genelists.txt')\n",
    "    df_seeds.columns = ['seeds']\n",
    "    df_seeds_z = pd.read_csv('input/seeds_from_zNumbers.txt', sep='\\n')\n",
    "    seeds_list = [str(i) for i in list(df_seeds['seeds'])]\n",
    "\n",
    "    d_seeds_idsym = {}\n",
    "    for k,v in d_gene_sym.items():\n",
    "        for i in seeds_list:\n",
    "            if v == i:\n",
    "                d_seeds_idsym[k]=i\n",
    "    seed_list = list(d_seeds_idsym.keys())\n",
    "\n",
    "    FM_BP = pd.read_csv('input/Features_GO_BiolProc_Dataframe_human.csv', index_col=0)\n",
    "    bioproc_list = list(FM_BP.index)\n",
    "\n",
    "    # From obtained Gene lists select those to be mapped onto different shells \n",
    "    # select genes for shells \n",
    "    shell_one = variant_list\n",
    "    shell_two = seed_list\n",
    "    shell_three = [str(i) for i in bioproc_list]\n",
    "\n",
    "    small_rad = 1\n",
    "    mid_rad = 5\n",
    "    large_rad = 30\n",
    "    outershell_rad = 50\n",
    "\n",
    "    d_param = {}\n",
    "    for i in G.nodes():\n",
    "        if str(i) in variant_list:\n",
    "            d_param[str(i)] = small_rad\n",
    "        elif i in seed_list:\n",
    "            d_param[str(i)] = mid_rad \n",
    "        elif int(i) in bioproc_list:\n",
    "            d_param[str(i)] = large_rad \n",
    "        else:\n",
    "            d_param[str(i)] = outershell_rad\n",
    "\n",
    "    d_radius = d_param\n",
    "    genes_rad = variant_list + seed_list + bioproc_list\n",
    "\n",
    "    # check how many shells of spheres there will be :\n",
    "    print('Number of Spherical Shells:', len(set(d_param.values())))\n",
    "\n",
    "\n",
    "    # Specifying coloring based on genes on different shells \n",
    "    d_nodecol = d_clos \n",
    "    d_colors = color_nodes_from_dict(G, d_nodecol, palette = col_pal)\n",
    "\n",
    "    d_colors_spec_genes = {}\n",
    "    for k,v in d_colors.items():\n",
    "        if k in shell_one:\n",
    "            d_colors_spec_genes[k]='#8b0000' \n",
    "        elif k in shell_two:\n",
    "            d_colors_spec_genes[k]='#FF4500' \n",
    "        elif k in shell_three:\n",
    "            d_colors_spec_genes[k]=v\n",
    "        else: \n",
    "            d_colors_spec_genes[k]='#d3d3d3'\n",
    "\n",
    "    d_colors_spec_genes_sort = {key:d_colors_spec_genes[key] for key in G.nodes()}\n",
    "    colors = list(d_colors_spec_genes_sort.values())\n",
    "\n",
    "    \n",
    "elif r_feat == 'essentiality':\n",
    "\n",
    "    small_rad = 1\n",
    "    mid_rad = 10\n",
    "    outershell_rad = 20\n",
    "\n",
    "    d_ess_scores = {}\n",
    "    for i in G.nodes():\n",
    "        if i in essential_genes:\n",
    "            d_ess_scores[i] = small_rad\n",
    "        elif i in non_ess_genes:\n",
    "            d_ess_scores[i] = mid_rad\n",
    "        elif str(i) in notdefined_genes:\n",
    "            d_ess_scores[i] = outershell_rad\n",
    "\n",
    "    d_radius = d_ess_scores\n",
    "    genes_rad = list(G.nodes())\n",
    "\n",
    "    # check how many shells of spheres there will be :\n",
    "    print('Number of Spherical Shells:', len(set(d_param.values())))\n",
    "\n",
    "    # Specifying coloring based on genes on different shells \n",
    "    d_colors_spec_genes = {}\n",
    "    for k in G.nodes():\n",
    "        if k in essential_genes:\n",
    "            d_colors_spec_genes[k]='#0080bf' \n",
    "        elif k in non_ess_genes:\n",
    "            d_colors_spec_genes[k]='#95dcff' \n",
    "        else: \n",
    "            d_colors_spec_genes[k]='#d3d3d3'\n",
    "\n",
    "    d_colors_spec_genes_sort = {key:d_colors_spec_genes[key] for key in G.nodes()}\n",
    "    colors = list(d_colors_spec_genes_sort.values())\n",
    "    \n",
    "    \n",
    "else: \n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "piano-devil",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### PLOT "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "violent-norway",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DM precalc used\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'output_plots/Geodesic/geodesic_r_r-autocore_local_human_light.html'"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ---------------------------------------------------------------------------\n",
    "#\n",
    "# SELECT a r - Parameter:\n",
    "#\n",
    "# d_radius > a dictionary with keys=G.nodes and values=any radius assigned to each node\n",
    "# ---------------------------------------------------------------------------\n",
    "\n",
    "posG_sphere = layout_geodesic(G, d_radius, n_neighbors=20, spread=1, min_dist=0.0)#, DM=DM)\n",
    "posG = posG_sphere\n",
    "\n",
    "umap_nodes = get_trace_nodes_3D(posG, l_features, colors, size3d, opacity_nodes)\n",
    "umap_edges = get_trace_edges_3D(G, posG, edge_color, opac = 0.05, linewidth = 0.5)\n",
    "data = [umap_nodes]#,umap_edges]\n",
    "\n",
    "path = 'output_plots/Geodesic/'\n",
    "schema = 'light' #'dark'\n",
    "fname = 'geodesic_r_'+r_feat+'_'+netlayout+'_'+organism+'_'+schema\n",
    "\n",
    "plot_3D(data,path,fname, 'dark')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:miniconda3-myenv_clip2]",
   "language": "python",
   "name": "conda-env-miniconda3-myenv_clip2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
