{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "instrumental-england",
   "metadata": {},
   "source": [
    "-----\n",
    "# cartoGRAPHs\n",
    "precalculated large networks , e.g. the Interactome\n",
    "\n",
    "A Notebook to produce 2D and 3D network layouts from any Graph,\n",
    "including interactive visualization (html files) and export functions \n",
    "to import into the VRNetzer analytics platform by Pirch et al.\n",
    "\n",
    "-----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "broadband-bacon",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cartoGRAPHs import * \n",
    "\n",
    "from func_load_data import *\n",
    "from func_visual_properties import * \n",
    "from func_calculations import * \n",
    "from func_embed_plot import * \n",
    "from func_exportVR import * "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unsigned-polish",
   "metadata": {},
   "source": [
    "_____\n",
    "# HUMAN NETWORK\n",
    "_____"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "romantic-chess",
   "metadata": {},
   "outputs": [],
   "source": [
    "organism = 'human'\n",
    "\n",
    "G = load_graph(organism) \n",
    "\n",
    "d_gene_sym = load_genesymbols(G, organism)\n",
    "l_features = list(d_gene_sym.values())\n",
    "\n",
    "d_gene_do = pickle.load( open( \"input/d_gene_do.pkl\", \"rb\" ) )\n",
    "d_do_genes = pickle.load( open( \"input/d_do_genes.pkl\", \"rb\" ) )\n",
    "d_do_names = pickle.load( open( \"input/DO_names.pkl\", \"rb\" ) )\n",
    "d_names_do = {y:x for x,y in d_do_names.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "uniform-romance",
   "metadata": {
    "tags": []
   },
   "source": [
    "_____ \n",
    "# 2 | FIGURE 2A + supplementary | Importance layout (with diverse node lists highlighted)\n",
    "_____"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "shaped-multiple",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_centralities = load_centralities(G, organism)\n",
    "df_centralities = pd.DataFrame(d_centralities).T\n",
    "df_centralities.columns = ['degree','closeness', 'betweeness', 'eigenvector']\n",
    "essential_genes,non_ess_genes,notdefined_genes = load_essentiality(G, organism)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "planned-laser",
   "metadata": {},
   "outputs": [],
   "source": [
    "DM_cent = load_datamatrix(G,organism, 'importance')\n",
    "DM_cent_mod = DM_cent.round(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "former-visitor",
   "metadata": {},
   "source": [
    "### 2D PORTRAIT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "fitted-worthy",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_n = 20\n",
    "spr = 1.0\n",
    "md = 0.0\n",
    "metr = 'cosine'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dental-respect",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time \n",
    "\n",
    "posG_umap2D = layout_portrait_umap(G,DM_cent_mod, 2,\n",
    "                                   n_neighbors = n_n,\n",
    "                                   spread = spr,\n",
    "                                   min_dist = md,\n",
    "                                   metric = metr) \n",
    "posG = posG_umap2D"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "exterior-order",
   "metadata": {
    "tags": []
   },
   "source": [
    "## VISUAL SETTINGS + PLOT PREPARATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unexpected-evans",
   "metadata": {},
   "source": [
    "#### NODES - GENERAL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "blessed-genealogy",
   "metadata": {},
   "outputs": [],
   "source": [
    "opacity_nodes = 0.8\n",
    "node_edge_col = '#696969' \n",
    "node_linewidth = 1\n",
    "\n",
    "scale_factor = 0.5\n",
    "#size_nx = list(draw_node_degree(G, scale_factor).values())\n",
    "size_plotly = list(draw_node_degree(G, scale_factor/100).values())\n",
    "\n",
    "scale_factor3D = 0.025\n",
    "size3d = list(draw_node_degree_3D(G, scale_factor3D).values())\n",
    "\n",
    "nodesglow_diameter = 8.0\n",
    "nodesglow_transparency = 0.01 # 0.01"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "arranged-thunder",
   "metadata": {},
   "source": [
    "#### ESSENTIALITY color settings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "attached-strain",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "color_method = 'essentiality'\n",
    "\n",
    "l_genes = essential_genes\n",
    "    \n",
    "node_col = '#00abff' \n",
    "undefined_col = '#DEDEDE '\n",
    "rest_col_nodes = '#DEDEDE '\n",
    "\n",
    "d_col_all = color_nodes_from_list(G, essential_genes, node_col)\n",
    "colors = list(d_col_all.values())\n",
    "    \n",
    "edge_color = node_col \n",
    "d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "width_edges = 0.25\n",
    "opacity_edges = 0.25"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caroline-blood",
   "metadata": {},
   "source": [
    "#### DEVELOPMENTAL color settings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "historical-recording",
   "metadata": {},
   "outputs": [],
   "source": [
    "color_method = 'oocyte'\n",
    "\n",
    "if color_method == 'oocyte' or color_method == '2cell' or color_method == '4cell' or color_method == '8cell':\n",
    "\n",
    "    node_col = '#0181A3' \n",
    "    devgenes_selection = color_method\n",
    "    df_devgenes = pd.read_csv('input/EmExplorer/'+devgenes_selection+'_Homo-sapiens.txt', delimiter = \"\\t\")\n",
    "    \n",
    "    devgenes = list(df_devgenes['stage'])\n",
    "\n",
    "    d_devgenes_entrezid = {}\n",
    "    for i,k in d_gene_sym.items():\n",
    "        if k in devgenes:\n",
    "             d_devgenes_entrezid[k] = i \n",
    "    devgenes_entrezid = list(d_devgenes_entrezid.values())\n",
    "    \n",
    "    l_genes = devgenes_entrezid\n",
    "    \n",
    "    d_col_all = color_nodes_from_list(G, l_genes, node_col)\n",
    "    colors = list(d_col_all.values())\n",
    "\n",
    "    edge_color = node_col\n",
    "    d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "    width_edges = 0.25\n",
    "    opacity_edges = 0.5\n",
    "        \n",
    "else:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stable-consensus",
   "metadata": {},
   "source": [
    "#### CANCER color settings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "commercial-expense",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_category = 'cancer'\n",
    "\n",
    "color_method = disease_category\n",
    "disease_genes = get_disease_genes(G, d_names_do, d_do_genes, disease_category)\n",
    "l_genes = disease_genes\n",
    "\n",
    "node_col = '#00C9FF' \n",
    "d_col_all = color_nodes_from_list(G, disease_genes, node_col)\n",
    "colors = list(d_col_all.values())\n",
    "\n",
    "edge_color = node_col \n",
    "d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "\n",
    "width_edges = 0.25\n",
    "opacity_edges = 0.25"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "rational-biotechnology",
   "metadata": {},
   "source": [
    "#### RARE DISEASE genes color settings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "gentle-queue",
   "metadata": {},
   "outputs": [],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "tree = ET.parse('input/orphadata.xml')\n",
    "root = tree.getroot()\n",
    "\n",
    "# retrieve genes from all rare diseases from database/file  \n",
    "disorder_to_genes = {}\n",
    "for disorder in root.iter('Disorder'):\n",
    "    dis = list(disorder.attrib.values())[0]\n",
    "    for elem in disorder.findall('DisorderGeneAssociationList'):\n",
    "        genes = []\n",
    "        for disorderassoc in elem.findall('DisorderGeneAssociation'):\n",
    "            for gene in disorderassoc.findall('Gene'):\n",
    "                for reference in gene.findall('ExternalReferenceList'):\n",
    "                     for refs in reference.findall('ExternalReference'):\n",
    "                        for ensg in refs.findall('Source'):\n",
    "                            if ensg.text == 'Ensembl':\n",
    "                                for e in refs.findall('Reference'):\n",
    "                                    ensgID = (e.text)\n",
    "                            else:\n",
    "                                pass\n",
    "                            \n",
    "            genes.append(ensgID)\n",
    "    disorder_to_genes[dis] = genes\n",
    "\n",
    "    \n",
    "# get a list of all rare disease genes obtained \n",
    "all_raredisease_genes = []\n",
    "for dis,genes in disorder_to_genes.items():\n",
    "    for i in genes:\n",
    "        all_raredisease_genes.append(i)\n",
    "      \n",
    "    \n",
    "# make sure there are no duplicates\n",
    "set_all_raredisease_genes = set(all_raredisease_genes)\n",
    "\n",
    "\n",
    "# convert ensgID to entrezID and get gene symbols\n",
    "ensg_to_entrez = pd.read_csv('input/ensg_to_entrezid.txt',sep=\"\t\")\n",
    "\n",
    "ensg = list(ensg_to_entrez['From'])\n",
    "entrez = [str(int(i)) for i in ensg_to_entrez['To'][:-1]] #list(ensg_to_entrez['To'])\n",
    "d_ensg_entrez = dict(zip(ensg,entrez))\n",
    "\n",
    "\n",
    "# match ensgID with entrezID\n",
    "set_all_raredisease_genes_id = {}\n",
    "for ix,entrez in d_ensg_entrez.items():\n",
    "    if ix in set_all_raredisease_genes:\n",
    "        set_all_raredisease_genes_id[ix]=entrez\n",
    "\n",
    "        \n",
    "# color all rare disease genes  \n",
    "l_rarediseasegenes = list(set_all_raredisease_genes_id.values())\n",
    "color_method = 'rarediseasegenes'\n",
    "\n",
    "l_genes = l_rarediseasegenes\n",
    "\n",
    "node_col =  '#004EAB'#'#0A81A1' \n",
    "undefined_col = '#DEDEDE '\n",
    "rest_col_nodes = '#DEDEDE '\n",
    "\n",
    "d_col_all = color_nodes_from_list(G, l_genes, node_col)\n",
    "colors = list(d_col_all.values())\n",
    "    \n",
    "edge_color = node_col \n",
    "d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "width_edges = 0.25\n",
    "opacity_edges = 0.25"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "accessory-martial",
   "metadata": {
    "tags": []
   },
   "source": [
    "## PLOT NETWORK PORTRAIT - plotly "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "attractive-february",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# plot nodes based on focus > separate to background/foreground \n",
    "\n",
    "posG_foreground = {}\n",
    "posG_background = {}\n",
    "for k,v in posG.items():\n",
    "    if k in l_genes:\n",
    "        posG_foreground[k]=v\n",
    "    else: \n",
    "        posG_background[k]=v\n",
    "    \n",
    "d_colors_foreground = {}\n",
    "d_colors_background = {}\n",
    "for i,c in d_col_all.items():\n",
    "    if i in posG_foreground.keys():\n",
    "        d_colors_foreground[i]=c\n",
    "    else: \n",
    "        d_colors_background[i]=c\n",
    "\n",
    "colors_foreground = list(d_colors_foreground.values())\n",
    "colors_background = list(d_colors_background.values())\n",
    "\n",
    "d_feat_foreground = {}\n",
    "d_feat_background = {}\n",
    "for k,v in d_gene_sym.items():\n",
    "    if k in posG_foreground: \n",
    "        d_feat_foreground[k]=v\n",
    "    else:\n",
    "        d_feat_background[k]=v\n",
    "        \n",
    "feat_foreground = list(d_feat_foreground.values())\n",
    "feat_background = list(d_feat_background.values())\n",
    "\n",
    "d_size_plotly = draw_node_degree(G, scale_factor/100)\n",
    "d_size_plotly_foreground = {}\n",
    "d_size_plotly_background = {}\n",
    "for k,v in d_size_plotly.items():\n",
    "    if k in posG_foreground.keys():\n",
    "        d_size_plotly_foreground[k]=v\n",
    "    else:\n",
    "        d_size_plotly_background[k]=v\n",
    "        \n",
    "size_plotly_foreground = list(d_size_plotly_foreground.values())\n",
    "size_plotly_background = list(d_size_plotly_background.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "sweet-pollution",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'output_plots/2Dlayouts/2Dportrait_NetlayoutImportance_rarediseasegenes_human_onlynodes.html'"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "umap_nodes_foreground = get_trace_nodes_2D(posG_foreground, feat_foreground, colors_foreground, size_plotly_foreground, \n",
    "                                           node_linewidth)\n",
    "umap_nodes_background = get_trace_nodes_2D(posG_background, feat_background, colors_background, size_plotly_background, \n",
    "                                           node_linewidth*0.25)\n",
    "umap_nodes_background_ = get_trace_nodes_2D(posG_background, feat_background, colors_background, 0.5, node_linewidth*0.5)\n",
    "\n",
    "umap_edges = get_trace_edges_specific2D(d_edge_col, posG, linew=width_edges, opac=opacity_edges)\n",
    "data = [umap_edges, \n",
    "    umap_nodes_background_, umap_nodes_background, umap_nodes_foreground]\n",
    "\n",
    "path = 'output_plots/2Dlayouts/'\n",
    "schema = 'light' #'dark' \n",
    "fname = '2Dportrait_NetlayoutImportance_'+color_method+'_'+organism\n",
    "\n",
    "plot_2D(data,path,fname)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faced-surge",
   "metadata": {},
   "source": [
    "## PLOT NETWORK PORTRAIT - networkx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "turned-disability",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Choose edges to highlight that link inbetween 2 essential genes only \n",
    "d_edge_col = color_edges_from_nodelist_specific(G, essential_genes, node_col)\n",
    "\n",
    "# Rest edges \n",
    "d_rest_edges={}\n",
    "for e in G.edges():\n",
    "    if e not in d_edge_col.keys():\n",
    "        d_rest_edges[e] = '#f0f0f0'\n",
    "\n",
    "d_all_edges = {**d_edge_col, **d_rest_edges}\n",
    "d_all_edges_sort = {key:d_all_edges[key] for key in G.edges()}\n",
    "edge_color = list(d_all_edges_sort.values())\n",
    "\n",
    "\n",
    "\n",
    "plt.figure(figsize=(25,25))\n",
    "plt.title('2D Portrait', size=16)\n",
    "\n",
    "nx.draw_networkx_nodes(G, \n",
    "                       posG,\n",
    "                       edgecolors = 'dimgrey', \n",
    "                       linewidths = 1.2, \n",
    "                       node_color = colors, \n",
    "                       node_size = size_nx, \n",
    "                       alpha = 0.8)\n",
    "\n",
    "nx.draw_networkx_edges(G, posG, width = 0.1, \n",
    "                       edge_color = edge_color, \n",
    "                      alpha = 0.25)\n",
    "\n",
    "plt.box(False)\n",
    "\n",
    "fname = 'output_plots/2Dlayouts/2Dportrait_Fig2a_'+organism\n",
    "\n",
    "# PNG\n",
    "#plt.savefig(fname+'.png',dpi=300)\n",
    "\n",
    "# PDF \n",
    "plt.savefig(fname+'.pdf',format='pdf')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "configured-spectrum",
   "metadata": {
    "tags": []
   },
   "source": [
    "_____ \n",
    "# FIGURE 2B - disease layout with ENHANCING FACTOR \n",
    "_____"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "brutal-group",
   "metadata": {
    "tags": []
   },
   "source": [
    "_____ \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "artificial-shelf",
   "metadata": {},
   "source": [
    "# 1. | NF1 + neighbors "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "peaceful-asset",
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_structural = load_datamatrix(G,organism, 'global')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "id": "hollow-dress",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENHANCING FACTOR functional features:  1.031327\n"
     ]
    }
   ],
   "source": [
    "# --------------------------------------\n",
    "#\n",
    "scalar_val = 11\n",
    "#\n",
    "# --------------------------------------\n",
    "\n",
    "df_max = DF_structural.max()\n",
    "l_max_visprob = max(list(df_max.values))\n",
    "\n",
    "enhance_factor = round((1-l_max_visprob)*scalar_val,10) # for emphasizing functional features > should be >= max of structural matrix values  # if significantly higher than max values in structural matrix > causes isolation of nodes with functional features enhanced\n",
    "print('ENHANCING FACTOR functional features: ',enhance_factor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "id": "invisible-democracy",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes: 50\n"
     ]
    }
   ],
   "source": [
    "path_nf = 'input/neurofibromatosis/'\n",
    "\n",
    "# gene of interest\n",
    "nf_goi_ = str(int(np.loadtxt(path_nf+'nf_gene_of_interest.txt', unpack=False)))\n",
    "nf_goi = list(nf_goi_.split(\" \"))\n",
    "nf_goi_str = ' '.join([str(item) for item in nf_goi])\n",
    "\n",
    "nf_neighbors = []\n",
    "for edge in G.edges():\n",
    "    if edge[0] == nf_goi_str:\n",
    "        nf_neighbors.append(edge[1])\n",
    "    elif edge[1] == nf_goi_str: \n",
    "        nf_neighbors.append(edge[0])\n",
    "\n",
    "enhance_genes = nf_goi + nf_neighbors # + nf_pathw + nf_neigh\n",
    "print('genes:', len(enhance_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "convenient-tucson",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Vector for NF1 + Neighbors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "id": "guilty-philippines",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>66008</th>\n",
       "      <th>8473</th>\n",
       "      <th>2561</th>\n",
       "      <th>3759</th>\n",
       "      <th>22906</th>\n",
       "      <th>4928</th>\n",
       "      <th>1994</th>\n",
       "      <th>8481</th>\n",
       "      <th>81610</th>\n",
       "      <th>51361</th>\n",
       "      <th>...</th>\n",
       "      <th>8001</th>\n",
       "      <th>51351</th>\n",
       "      <th>1551</th>\n",
       "      <th>51458</th>\n",
       "      <th>143903</th>\n",
       "      <th>10861</th>\n",
       "      <th>51471</th>\n",
       "      <th>221044</th>\n",
       "      <th>29965</th>\n",
       "      <th>NF1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66008</th>\n",
       "      <td>0.900242</td>\n",
       "      <td>0.000687</td>\n",
       "      <td>0.012865</td>\n",
       "      <td>0.006433</td>\n",
       "      <td>0.005347</td>\n",
       "      <td>0.000460</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>0.000345</td>\n",
       "      <td>0.003464</td>\n",
       "      <td>0.002905</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8473</th>\n",
       "      <td>0.002519</td>\n",
       "      <td>0.900245</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.005316</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2561</th>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.900367</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3759</th>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.900389</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22906</th>\n",
       "      <td>0.002525</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.900120</td>\n",
       "      <td>0.000459</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143903</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.900070</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10861</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.900070</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51471</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.900052</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221044</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.90003</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29965</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.900063</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16376 rows × 16377 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           66008      8473      2561      3759     22906      4928      1994  \\\n",
       "66008   0.900242  0.000687  0.012865  0.006433  0.005347  0.000460  0.000051   \n",
       "8473    0.002519  0.900245  0.000036  0.000018  0.005316  0.000005  0.000002   \n",
       "2561    0.002502  0.000002  0.900367  0.000018  0.000015  0.000002  0.000000   \n",
       "3759    0.002502  0.000002  0.000036  0.900389  0.000015  0.000002  0.000000   \n",
       "22906   0.002525  0.000685  0.000036  0.000018  0.900120  0.000459  0.000000   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "143903  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "10861   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "51471   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "221044  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "29965   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "            8481     81610     51361  ...      8001     51351  1551  51458  \\\n",
       "66008   0.000345  0.003464  0.002905  ...  0.000000  0.000000   0.0    0.0   \n",
       "8473    0.000003  0.000012  0.000008  ...  0.000000  0.000000   0.0    0.0   \n",
       "2561    0.000001  0.000010  0.000008  ...  0.000000  0.000001   0.0    0.0   \n",
       "3759    0.000001  0.000010  0.000008  ...  0.000001  0.000000   0.0    0.0   \n",
       "22906   0.000001  0.000010  0.000009  ...  0.000000  0.000000   0.0    0.0   \n",
       "...          ...       ...       ...  ...       ...       ...   ...    ...   \n",
       "143903  0.000000  0.000000  0.000000  ...  0.000000  0.000070   0.0    0.0   \n",
       "10861   0.000000  0.000000  0.000000  ...  0.000000  0.000070   0.0    0.0   \n",
       "51471   0.000000  0.000000  0.000000  ...  0.000000  0.000000   0.0    0.0   \n",
       "221044  0.000000  0.000000  0.000000  ...  0.000000  0.000000   0.0    0.0   \n",
       "29965   0.000000  0.000000  0.000000  ...  0.000000  0.000000   0.0    0.0   \n",
       "\n",
       "          143903     10861     51471   221044     29965  NF1  \n",
       "66008   0.000000  0.000000  0.000000  0.00000  0.000000  0.0  \n",
       "8473    0.000000  0.000000  0.000000  0.00000  0.000000  0.0  \n",
       "2561    0.000001  0.000001  0.000000  0.00000  0.000000  0.0  \n",
       "3759    0.000000  0.000000  0.000000  0.00000  0.000000  0.0  \n",
       "22906   0.000000  0.000000  0.000000  0.00000  0.000000  0.0  \n",
       "...          ...       ...       ...      ...       ...  ...  \n",
       "143903  0.900070  0.000070  0.000000  0.00000  0.000000  0.0  \n",
       "10861   0.000070  0.900070  0.000000  0.00000  0.000000  0.0  \n",
       "51471   0.000000  0.000000  0.900052  0.00000  0.000000  0.0  \n",
       "221044  0.000000  0.000000  0.000000  0.90003  0.000000  0.0  \n",
       "29965   0.000000  0.000000  0.000000  0.00000  0.900063  0.0  \n",
       "\n",
       "[16376 rows x 16377 columns]"
      ]
     },
     "execution_count": 457,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d_vec_disease = {}\n",
    "for i in G.nodes():\n",
    "    if i in enhance_genes:\n",
    "        d_vec_disease[i]=1\n",
    "    else:\n",
    "        d_vec_disease[i]=0\n",
    "        \n",
    "d_vec_disease_sorted = {key:d_vec_disease[key] for key in G.nodes()}\n",
    "DF_nf = pd.DataFrame(list(d_vec_disease_sorted.values()))\n",
    "DF_nf.index = list(G.nodes())\n",
    "DF_nf.columns = ['NF1']\n",
    "\n",
    "DF_nf = DF_nf.T\n",
    "DF_structural.index = DF_structural.columns\n",
    "\n",
    "DF_nf_scal = DF_nf * enhance_factor\n",
    "DF_merge_nf = pd.concat([DF_structural, \n",
    "                      DF_nf_scal.T],axis=1)\n",
    "\n",
    "DF_merge_nf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "independent-concert",
   "metadata": {},
   "source": [
    "### 2D PORTRAIT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "id": "temporal-lingerie",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_n = 8\n",
    "spr = 1.0\n",
    "md = 0.0\n",
    "metr = 'cosine'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "id": "developmental-migration",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 51s, sys: 10.4 s, total: 2min 2s\n",
      "Wall time: 1min 20s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "posG_umap2D = layout_portrait_umap(G,DF_merge_nf, 2,\n",
    "                                   n_neighbors = n_n,\n",
    "                                   spread = spr,\n",
    "                                   min_dist = md,\n",
    "                                   metric = metr) \n",
    "posG = posG_umap2D"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "contained-funeral",
   "metadata": {
    "tags": []
   },
   "source": [
    "## VISUAL SETTINGS + PLOT PREPARATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "historic-coverage",
   "metadata": {},
   "source": [
    "#### NODES - GENERAL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "id": "pretty-journalism",
   "metadata": {},
   "outputs": [],
   "source": [
    "opacity_nodes = 0.8\n",
    "node_edge_col = '#696969' \n",
    "node_linewidth = 1\n",
    "\n",
    "scale_factor = 0.5\n",
    "size_plotly = list(draw_node_degree(G, scale_factor/100).values())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "anonymous-pledge",
   "metadata": {},
   "source": [
    "#### NF1 SPECIFIC "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 461,
   "id": "invisible-scroll",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAecAAAEzCAYAAAALosttAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAEQlJREFUeJzt3W2MnNdZx+H/jU0QlJYCNajkhRqaUvyBQmpCERSKCiWOAPMmlIBoKVRWpAaBEFKDKgESn0oFQqgBy0BUioAgRAEXGQJC0H6AQJySpjEhrZu+2CQ0bot4KyK4vfkwj2Gy7HpnnXFyZue6pJVnzpyZPUfPJj/Ps+tnq7sDAIzjU57uBQAATyTOADAYcQaAwYgzAAxGnAFgMOIMAIPZNs5VdUdVPVZVD2zxeFXVL1bV6aq6v6quW/4yAWB9LPLO+c1JbrjI44eSXDt9HEnyy09+WQCwvraNc3e/I8nHLjLlcJK39MzdSZ5dVc9d1gIBYN0s43vOVyY5M3f/7DQGAFyCvUt4jdpkbNNrglbVkcxOfecZz3jGi1/4whc+6U/++OOPLzTvUi9TugrPW4U1XurzVmGNl/q8nTxnFfbzVD9vFdb4VD9vFdZ4qc97Kj7XpX6O5z//+QvPvffeez/S3fu2m7eMOJ9NcvXc/auSPLLZxO4+luRYkhw8eLBPnjz5pD/5mTNnLrz2/OfZcmwZc5+Kz2Guuas8d7T1rMvazX3q5s5729vetun4Zqrqg4vMW8Zp7eNJXjn91PZLkvxLdz+6hNcFgLW07TvnqvrtJC9L8pyqOpvkp5J8apJ099EkJ5LcmOR0ko8nefXlWiwArINt49zdN2/zeCd57dJWBABrzhXCAGAw4gwAgxFnABiMOAPAYMQZAAYjzgAwGHEGgMGIMwAMRpwBYDDiDACDEWcAGIw4A8BgxBkABiPOADAYcQaAwYgzAAxGnAFgMOIMAIMRZwAYjDgDwGDEGQAGI84AMBhxBoDBiDMADEacAWAw4gwAgxFnABiMOAPAYMQZAAYjzgAwGHEGgMGIMwAMRpwBYDDiDACDEWcAGIw4A8BgxBkABiPOADAYcQaAwYgzAAxGnAFgMOIMAIMRZwAYjDgDwGDEGQAGI84AMJiF4lxVN1TVQ1V1uqpu2+Txz6qqt1XVu6rqVFW9evlLBYD1sG2cq2pPktuTHEpyIMnNVXVgw7TXJvn77n5Rkpcl+bmqumLJawWAtbDIO+frk5zu7oe7+/EkdyY5vGFOJ3lmVVWSz0zysSTnl7pSAFgTi8T5yiRn5u6fncbmvSnJlyZ5JMm7k/xId39y4wtV1ZGqOllVJ8+dO3eJSwaA3W2RONcmY73h/jcnuS/JFyT58iRvqqpn/b8ndR/r7oPdfXDfvn07XiwArINF4nw2ydVz96/K7B3yvFcneWvPnE7y/iQvXM4SAWC9LBLne5JcW1X7px/yuinJ8Q1zPpTk5UlSVZ+f5EuSPLzMhQLAuti73YTuPl9Vtya5K8meJHd096mqumV6/GiSn0ny5qp6d2anwV/X3R+5jOsGgF1r2zgnSXefSHJiw9jRuduPJHnFcpcGAOvJFcIAYDDiDACDEWcAGIw4A8BgxBkABiPOADAYcQaAwYgzAAxGnAFgMOIMAIMRZwAYjDgDwGDEGQAGI84AMBhxBoDBiDMADEacAWAw4gwAgxFnABiMOAPAYMQZAAYjzgAwGHEGgMGIMwAMRpwBYDDiDACDEWcAGIw4A8BgxBkABiPOADAYcQaAwYgzAAxGnAFgMOIMAIMRZwAYjDgDwGDEGQAGI84AMBhxBoDBiDMADEacAWAw4gwAgxFnABiMOAPAYMQZAAazUJyr6oaqeqiqTlfVbVvMeVlV3VdVp6rq7ctdJgCsj73bTaiqPUluT/JNSc4muaeqjnf338/NeXaSX0pyQ3d/qKo+73ItGAB2u0XeOV+f5HR3P9zdjye5M8nhDXO+N8lbu/tDSdLdjy13mQCwPhaJ85VJzszdPzuNzXtBks+uqr+sqnur6pXLWiAArJttT2snqU3GepPXeXGSlyf59CR/XVV3d/d7nvBCVUeSHEmSa665ZuerBYA1sMg757NJrp67f1WSRzaZ8yfd/R/d/ZEk70jyoo0v1N3Huvtgdx/ct2/fpa4ZAHa1ReJ8T5Jrq2p/VV2R5KYkxzfM+cMkL62qvVX1GUm+KsmDy10qAKyHbU9rd/f5qro1yV1J9iS5o7tPVdUt0+NHu/vBqvqTJPcn+WSSX+3uBy7nwgFgt1rke87p7hNJTmwYO7rh/huTvHF5SwOA9eQKYQAwGHEGgMGIMwAMRpwBYDDiDACDEWcAGIw4A8BgxBkABiPOADAYcQaAwYgzAAxGnAFgMOIMAIMRZwAYjDgDwGDEGQAGI84AMBhxBoDBiDMADEacAWAw4gwAgxFnABiMOAPAYMQZAAYjzgAwGHEGgMGIMwAMRpwBYDDiDACDEWcAGIw4A8BgxBkABiPOADAYcQaAwYgzAAxGnAFgMOIMAIMRZwAYjDgDwGDEGQAGI84AMBhxBoDBiDMADEacAWAw4gwAg1kozlV1Q1U9VFWnq+q2i8z7yqr6RFV99/KWCADrZds4V9WeJLcnOZTkQJKbq+rAFvPekOSuZS8SANbJIu+cr09yursf7u7Hk9yZ5PAm8344ye8leWyJ6wOAtbNInK9Mcmbu/tlp7H9V1ZVJviPJ0eUtDQDW0yJxrk3GesP9X0jyuu7+xEVfqOpIVZ2sqpPnzp1bdI0AsFb2LjDnbJKr5+5fleSRDXMOJrmzqpLkOUlurKrz3f0H85O6+1iSY0ly8ODBjYEHALJYnO9Jcm1V7U/yj0luSvK98xO6e/+F21X15iR/tDHMAMBito1zd5+vqlsz+ynsPUnu6O5TVXXL9LjvMwPAEi3yzjndfSLJiQ1jm0a5u3/gyS8LANaXK4QBwGDEGQAGI84AMBhxBoDBiDMADEacAWAw4gwAgxFnABiMOAPAYMQZAAYjzgAwGHEGgMGIMwAMRpwBYDDiDACDEWcAGIw4A8BgxBkABiPOADAYcQaAwYgzAAxGnAFgMOIMAIMRZwAYjDgDwGDEGQAGI84AMBhxBoDBiDMADEacAWAw4gwAgxFnABiMOAPAYMQZAAYjzgAwGHEGgMGIMwAMRpwBYDDiDACDEWcAGIw4A8BgxBkABiPOADAYcQaAwYgzAAxGnAFgMAvFuapuqKqHqup0Vd22yePfV1X3Tx9/VVUvWv5SAWA9bBvnqtqT5PYkh5IcSHJzVR3YMO39Sb6+u78syc8kObbshQLAuljknfP1SU5398Pd/XiSO5Mcnp/Q3X/V3f883b07yVXLXSYArI9F4nxlkjNz989OY1v5oSR/vNkDVXWkqk5W1clz584tvkoAWCOLxLk2GetNJ1Z9Q2Zxft1mj3f3se4+2N0H9+3bt/gqAWCN7F1gztkkV8/dvyrJIxsnVdWXJfnVJIe6+6PLWR4ArJ9F3jnfk+TaqtpfVVckuSnJ8fkJVXVNkrcm+f7ufs/ylwkA62Pbd87dfb6qbk1yV5I9Se7o7lNVdcv0+NEkP5nkc5P8UlUlyfnuPnj5lg0Au9cip7XT3SeSnNgwdnTu9muSvGa5SwOA9eQKYQAwGHEGgMGIMwAMRpwBYDDiDACDEWcAGIw4A8BgxBkABiPOADAYcQaAwYgzAAxGnAFgMOIMAIMRZwAYjDgDwGDEGQAGI84AMBhxBoDBiDMADEacAWAw4gwAgxFnABiMOAPAYMQZAAYjzgAwGHEGgMGIMwAMRpwBYDDiDACDEWcAGIw4A8BgxBkABiPOADAYcQaAwYgzAAxGnAFgMOIMAIMRZwAYjDgDwGDEGQAGI84AMBhxBoDBiDMADEacAWAw4gwAg1kozlV1Q1U9VFWnq+q2TR6vqvrF6fH7q+q65S8VANbDtnGuqj1Jbk9yKMmBJDdX1YEN0w4luXb6OJLkl5e8TgBYG4u8c74+yenufri7H09yZ5LDG+YcTvKWnrk7ybOr6rlLXisArIVF4nxlkjNz989OYzudAwAsYO8Cc2qTsb6EOamqI5md9k6Sf6+qhxb4/It6TpKPLPH1RmFfq8W+Vot9rZYh91W1WQK39IWLTFokzmeTXD13/6okj1zCnHT3sSTHFlnYTlXVye4+eDle++lkX6vFvlaLfa2W3bqvzSxyWvueJNdW1f6quiLJTUmOb5hzPMkrp5/afkmSf+nuR5e8VgBYC9u+c+7u81V1a5K7kuxJckd3n6qqW6bHjyY5keTGJKeTfDzJqy/fkgFgd1vktHa6+0RmAZ4fOzp3u5O8drlL27HLcrp8APa1WuxrtdjXatmt+/p/atZVAGAULt8JAIPZFXHe7vKiI6uqq6vqL6rqwao6VVU/Mo3/dFX9Y1XdN33cOPecn5j2+lBVffPTt/qLq6oPVNW7p/WfnMY+p6r+rKreO/352XPzh99XVX3J3DG5r6r+tap+dBWPV1XdUVWPVdUDc2M7Pj5V9eLpOJ+eLuO7o39Xsmxb7OuNVfUP0+WFf7+qnj2NP6+q/nPuuB2de84w+9piTzv+mhtpT9N6NtvX78zt6QNVdd80vhLHamm6e6U/Mvshtfcl+aIkVyR5V5IDT/e6drD+5ya5brr9zCTvyewyqT+d5Mc3mX9g2uOnJdk/7X3P072PLfb2gSTP2TD2s0lum27fluQNq7avDV97/5TZv1tcueOV5OuSXJfkgSdzfJL8bZKvzux6B3+c5NCA+3pFkr3T7TfM7et58/M2vM4w+9piTzv+mhtpT1vta8PjP5fkJ1fpWC3rYze8c17k8qLD6u5Hu/ud0+1/S/JgLn51tcNJ7uzu/+ru92f2E/LXX/6VLs3hJL8+3f71JN8+N75q+3p5kvd19wcvMmfYfXX3O5J8bMPwjo5PzS7T+6zu/uue/V/yLXPPeVpstq/u/tPuPj/dvTuzazFsabR9bXGstrLSx+qC6d3v9yT57Yu9xoj7WobdEOddc+nQqnpekq9I8jfT0K3Tabg75k4vrtJ+O8mfVtW9Nbs6XJJ8fk//Bn768/Om8VXa1wU35Yn/41j145Xs/PhcOd3eOD6yH8zs3dUF+6vq76rq7VX10mlsVfa1k6+5VdnTBS9N8uHufu/c2Cofqx3ZDXFe6NKho6uqz0zye0l+tLv/NbPf7PXFSb48yaOZnd5JVmu/X9Pd12X2W8teW1Vfd5G5q7Sv1OyCPN+W5Henod1wvC5mq32s1P6q6vVJzif5zWno0STXdPdXJPmxJL9VVc/Kauxrp19zq7CneTfniX/5XeVjtWO7Ic4LXTp0ZFX1qZmF+Te7+61J0t0f7u5PdPcnk/xK/u9U6Mrst7sfmf58LMnvZ7aHD0+noS6cjnpsmr4y+5ocSvLO7v5wsjuO12Snx+dsnniKeNj9VdWrknxLku+bTn9mOvX70en2vZl9f/YFWYF9XcLX3PB7uqCq9ib5ziS/c2FslY/VpdgNcV7k8qLDmr6v8mtJHuzun58bn/+Vm9+R5MJPMx5PclNVfVpV7c/sd2j/7VO13kVV1TOq6pkXbmf2AzkPZLb+V03TXpXkD6fbK7GvOU/4W/2qH685Ozo+06nvf6uql0xfy6+ce84wquqGJK9L8m3d/fG58X01+531qaovymxfD6/Cvnb6NbcKe5rzjUn+obv/93T1Kh+rS/J0/0TaMj4yu3ToezL7m9Trn+717HDtX5vZKZj7k9w3fdyY5DeSvHsaP57kuXPPef2014cy6E8lZvbT8++aPk5dOC5JPjfJnyd57/Tn56zSvqZ1fkaSjyb5rLmxlTtemf3l4tEk/53Zu48fupTjk+RgZmF4X5I3Zbq40WD7Op3Z92Ev/Dd2dJr7XdPX57uSvDPJt464ry32tOOvuZH2tNW+pvE3J7llw9yVOFbL+nCFMAAYzG44rQ0Au4o4A8BgxBkABiPOADAYcQaAwYgzAAxGnAFgMOIMAIP5H7VBmqrwzKarAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 576x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "color_method = 'NF1'\n",
    "l_genes = enhance_genes\n",
    "\n",
    "node_col = '#00E3DF' #'#FF8E00' \n",
    "undefined_col = '#DEDEDE'\n",
    "rest_col_nodes = '#DEDEDE'\n",
    "\n",
    "# color nodes\n",
    "col_param = list(df_centralities['closeness'].values)\n",
    "d_node_colors = dict(zip(list(G.nodes()), col_param))\n",
    "\n",
    "c1='#E2E2E2' #'dimgrey' #blue\n",
    "c2='#4C4C4C' #'#6A6A6A'  #'lightgrey' #green\n",
    "n=len(set(col_param))\n",
    "\n",
    "#fig, ax = plt.subplots(figsize=(8, 5))\n",
    "colors_for_palette = []\n",
    "for x in range(n+1):\n",
    "#   ax.axvline(x, color=colorFader(c1,c2,x/n), linewidth=4)\n",
    "    colors_for_palette.append(colorFader(c1,c2,x/n))    \n",
    "#plt.show()\n",
    "\n",
    "# color edges    \n",
    "edge_color = node_col \n",
    "d_edge_col = color_edges_from_node(G, nf_goi, node_col)\n",
    "width_edges = 0.7\n",
    "opacity_edges = 0.7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "id": "alien-thong",
   "metadata": {},
   "outputs": [],
   "source": [
    "#d_col_all = color_nodes_from_list(G, l_genes, node_col)\n",
    "#colors = list(d_col_all.values())\n",
    "\n",
    "customPalette = sns.set_palette(sns.color_palette(colors_for_palette))\n",
    "d_node_colors = dict(zip(list(G.nodes()), col_param))\n",
    "\n",
    "col_pal = customPalette\n",
    "\n",
    "d_colors = color_nodes_from_dict(G, d_node_colors, palette = col_pal)\n",
    "colors = list(d_colors.values())\n",
    "edge_color = '#d3d3d3'\n",
    "\n",
    "d_col_all = {}\n",
    "for k,v in d_colors.items():\n",
    "    if k in l_genes:\n",
    "        d_col_all[k]=node_col\n",
    "    else:\n",
    "        d_col_all[k]=v"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "blank-corpus",
   "metadata": {
    "tags": []
   },
   "source": [
    "### PLOT NETWORK PORTRAIT - plotly "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 463,
   "id": "inappropriate-casting",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# plot nodes based on focus > separate to background/foreground \n",
    "\n",
    "posG_foreground = {}\n",
    "posG_background = {}\n",
    "for k,v in posG.items():\n",
    "    if k in l_genes:\n",
    "        posG_foreground[k]=v\n",
    "    else: \n",
    "        posG_background[k]=v\n",
    "    \n",
    "d_colors_foreground = {}\n",
    "d_colors_background = {}\n",
    "for i,c in d_col_all.items():\n",
    "    if i in posG_foreground.keys():\n",
    "        d_colors_foreground[i]=c\n",
    "    else: \n",
    "        d_colors_background[i]=c\n",
    "\n",
    "colors_foreground = list(d_colors_foreground.values())\n",
    "colors_background = list(d_colors_background.values())\n",
    "\n",
    "d_feat_foreground = {}\n",
    "d_feat_background = {}\n",
    "for k,v in d_gene_sym.items():\n",
    "    if k in posG_foreground: \n",
    "        d_feat_foreground[k]=v\n",
    "    else:\n",
    "        d_feat_background[k]=v\n",
    "        \n",
    "feat_foreground = list(d_feat_foreground.values())\n",
    "feat_background = list(d_feat_background.values())\n",
    "\n",
    "d_size_plotly = draw_node_degree(G, scale_factor/100)\n",
    "d_size_plotly_foreground = {}\n",
    "d_size_plotly_background = {}\n",
    "for k,v in d_size_plotly.items():\n",
    "    if k in posG_foreground.keys():\n",
    "        d_size_plotly_foreground[k]=v\n",
    "    else:\n",
    "        d_size_plotly_background[k]=v\n",
    "        \n",
    "size_plotly_foreground = list(d_size_plotly_foreground.values())\n",
    "size_plotly_background = list(d_size_plotly_background.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "id": "matched-leonard",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'output_plots/2Dlayouts/2Dportrait_NetlayoutDisease_NF1Neighbors_enhance11_NF1_nn8_spr1.0_md0.0_cosine_human.html'"
      ]
     },
     "execution_count": 465,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "umap_nodes_foreground = get_trace_nodes_2D(posG_foreground, feat_foreground, colors_foreground, size_plotly_foreground, \n",
    "                                           node_linewidth, 0.8)\n",
    "umap_nodes_background = get_trace_nodes_2D(posG_background, feat_background, colors_background, size_plotly_background, \n",
    "                                           node_linewidth*0.25, 0.5)\n",
    "umap_nodes_background_ = get_trace_nodes_2D(posG_background, feat_background, colors_background, 2, \n",
    "                                            None,0.25)\n",
    "\n",
    "umap_edges = get_trace_edges_specific2D(d_edge_col, posG, linew=width_edges, opac=opacity_edges)\n",
    "data = [umap_edges, \n",
    "        umap_nodes_background_, \n",
    "        umap_nodes_background, \n",
    "        umap_nodes_foreground]\n",
    "\n",
    "path = 'output_plots/2Dlayouts/'\n",
    "schema = 'light' #'dark' \n",
    "fname = '2Dportrait_NetlayoutDisease_NF1Neighbors_enhance'+str(scalar_val)+'_'+color_method+'_nn'+str(n_n)+'_spr'+str(spr)+'_md'+str(md)+'_'+str(metr)+'_'+organism\n",
    "\n",
    "plot_2D(data,path,fname)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "south-adventure",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 2. | NF1 in Disease Module\n",
    "e.g.\n",
    "+ DOID:0111253 / Neurofibromatosis1\n",
    "+ DOID:8712 / Neurofibromatosis\n",
    "+ DOID:962 / Neurofibroma\n",
    "+ DOID:0050736 / Autosomal Dominant Disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 615,
   "id": "smoking-police",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENHANCING FACTOR first:  0.468785\n",
      "ENHANCING FACTOR second:  0.468785\n"
     ]
    }
   ],
   "source": [
    "# --------------------------------------\n",
    "#\n",
    "scalar_val_1 = 5\n",
    "scalar_val_2 = 5\n",
    "#\n",
    "# --------------------------------------\n",
    "\n",
    "enhance_factor_1 = round((1-l_max_visprob)*scalar_val_1,10)\n",
    "enhance_factor_2 = round((1-l_max_visprob)*scalar_val_2,10)\n",
    "print('ENHANCING FACTOR first: ',enhance_factor_1)\n",
    "print('ENHANCING FACTOR second: ',enhance_factor_2)\n",
    "\n",
    "n_n = 8 #6 #20\n",
    "spr = 1.0\n",
    "md = 0.0 #0.4 #0.4\n",
    "metr = 'cosine'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 616,
   "id": "smaller-curtis",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Genes associated to DOID 962 : 10\n",
      "All genes: 60\n"
     ]
    }
   ],
   "source": [
    "# get Disease Feature Matrix \n",
    "\n",
    "FM_disease = pd.read_csv('input/Features_Disease_Dataframe_'+organism+'.csv', index_col=0)\n",
    "\n",
    "num_doid = '962'\n",
    "\n",
    "# define specific DOID \n",
    "dismod = FM_disease.loc[:,'DOID:'+num_doid]\n",
    "dismod.index = FM_disease.index\n",
    "dismod_id = 'DOID'+num_doid\n",
    "\n",
    "dismod_genes = [] \n",
    "for i,v in enumerate(dismod.values):\n",
    "    if v == 1:\n",
    "        dismod_genes.append(dismod.index[i])\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "print('Number of Genes associated to DOID', num_doid,':',len(dismod_genes))\n",
    "\n",
    "\n",
    "path_nf = 'input/neurofibromatosis/'\n",
    "\n",
    "# gene of interest\n",
    "nf_goi_ = str(int(np.loadtxt(path_nf+'nf_gene_of_interest.txt', unpack=False)))\n",
    "nf_goi = list(nf_goi_.split(\" \"))\n",
    "nf_goi_str = ' '.join([str(item) for item in nf_goi])\n",
    "\n",
    "nf_neighbors = []\n",
    "for edge in G.edges():\n",
    "    if edge[0] == nf_goi_str:\n",
    "        nf_neighbors.append(edge[1])\n",
    "    elif edge[1] == nf_goi_str: \n",
    "        nf_neighbors.append(edge[0])\n",
    "\n",
    "enhance_genes = nf_goi + nf_neighbors\n",
    "l_genes = [str(i) for i in dismod_genes] + enhance_genes \n",
    "\n",
    "print('All genes:', len(l_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "governmental-ranking",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Vector for Disease Module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 617,
   "id": "funded-leave",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>66008</th>\n",
       "      <th>8473</th>\n",
       "      <th>2561</th>\n",
       "      <th>3759</th>\n",
       "      <th>22906</th>\n",
       "      <th>4928</th>\n",
       "      <th>1994</th>\n",
       "      <th>8481</th>\n",
       "      <th>81610</th>\n",
       "      <th>51361</th>\n",
       "      <th>...</th>\n",
       "      <th>51351</th>\n",
       "      <th>1551</th>\n",
       "      <th>51458</th>\n",
       "      <th>143903</th>\n",
       "      <th>10861</th>\n",
       "      <th>51471</th>\n",
       "      <th>221044</th>\n",
       "      <th>29965</th>\n",
       "      <th>NF1</th>\n",
       "      <th>Disease Module</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66008</th>\n",
       "      <td>0.900242</td>\n",
       "      <td>0.002519</td>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.002525</td>\n",
       "      <td>0.002517</td>\n",
       "      <td>0.002512</td>\n",
       "      <td>0.002511</td>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.002501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8473</th>\n",
       "      <td>0.000687</td>\n",
       "      <td>0.900245</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2561</th>\n",
       "      <td>0.012865</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.900367</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.468785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3759</th>\n",
       "      <td>0.006433</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.900389</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.468785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22906</th>\n",
       "      <td>0.005347</td>\n",
       "      <td>0.005316</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.900120</td>\n",
       "      <td>0.005314</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.468785</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 16378 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          66008      8473      2561      3759     22906      4928      1994  \\\n",
       "66008  0.900242  0.002519  0.002502  0.002502  0.002525  0.002517  0.002512   \n",
       "8473   0.000687  0.900245  0.000002  0.000002  0.000685  0.000008  0.000031   \n",
       "2561   0.012865  0.000036  0.900367  0.000036  0.000036  0.000065  0.000037   \n",
       "3759   0.006433  0.000018  0.000018  0.900389  0.000018  0.000027  0.000041   \n",
       "22906  0.005347  0.005316  0.000015  0.000015  0.900120  0.005314  0.000026   \n",
       "\n",
       "           8481     81610     51361  ...  51351  1551  51458  143903  10861  \\\n",
       "66008  0.002511  0.002502  0.002501  ...    0.0   0.0    0.0     0.0    0.0   \n",
       "8473   0.000006  0.000002  0.000002  ...    0.0   0.0    0.0     0.0    0.0   \n",
       "2561   0.000036  0.000036  0.000036  ...    0.0   0.0    0.0     0.0    0.0   \n",
       "3759   0.000018  0.000018  0.000018  ...    0.0   0.0    0.0     0.0    0.0   \n",
       "22906  0.000015  0.000015  0.000016  ...    0.0   0.0    0.0     0.0    0.0   \n",
       "\n",
       "       51471  221044  29965  NF1  Disease Module  \n",
       "66008    0.0     0.0    0.0  0.0        0.000000  \n",
       "8473     0.0     0.0    0.0  0.0        0.000000  \n",
       "2561     0.0     0.0    0.0  0.0        0.468785  \n",
       "3759     0.0     0.0    0.0  0.0        0.468785  \n",
       "22906    0.0     0.0    0.0  0.0        0.468785  \n",
       "\n",
       "[5 rows x 16378 columns]"
      ]
     },
     "execution_count": 617,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create an empty matrix for zeros of rest genes (not associated to any disease)\n",
    "rest = []\n",
    "for i in G.nodes():\n",
    "    if str(i) in dismod.index or int(i) in dismod.index:\n",
    "        pass\n",
    "    else: \n",
    "        rest.append(i)\n",
    "        \n",
    "empty = np.zeros(shape=(len(rest),1))\n",
    "empty_stacked = np.vstack(empty)\n",
    "df_empty = pd.DataFrame(empty_stacked)\n",
    "df_empty.index = rest\n",
    "\n",
    "# combine disease vector with rest of genes in graph\n",
    "df_dismod_complete = pd.concat([dismod, df_empty])\n",
    "\n",
    "# reorder according to G.nodes\n",
    "df_dismod_complete.columns = ['Disease Module']\n",
    "DF_dismod_reorder = df_dismod_complete.reindex(list(G.nodes()))\n",
    "\n",
    "# fill nan positions (due to reordering) with 1 \n",
    "DF_dismod_new = DF_dismod_reorder.replace(np.nan,1)\n",
    "\n",
    "d_vec_disease = {}\n",
    "for i in G.nodes():\n",
    "    if i in l_genes:\n",
    "        d_vec_disease[i]=1\n",
    "    else:\n",
    "        d_vec_disease[i]=0\n",
    "        \n",
    "d_vec_disease_sorted = {key:d_vec_disease[key] for key in G.nodes()}\n",
    "DF_nf = pd.DataFrame(list(d_vec_disease_sorted.values()))\n",
    "DF_nf.index = list(G.nodes())\n",
    "DF_nf.columns = ['NF1']\n",
    "\n",
    "DF_nf = DF_nf.T\n",
    "DF_structural.index = DF_structural.columns\n",
    "\n",
    "# multiply with enhancing factor\n",
    "DF_nf_scal = DF_nf * enhance_factor_1\n",
    "DF_dismod_scal = DF_dismod_new * enhance_factor_2\n",
    "\n",
    "DF_merge_dismod = pd.concat([DF_structural.T, \n",
    "                             DF_nf_scal.T,\n",
    "                             DF_dismod_scal],axis=1)\n",
    "\n",
    "DF_merge_dismod.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 618,
   "id": "impossible-peace",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2min 16s, sys: 13.1 s, total: 2min 29s\n",
      "Wall time: 1min 33s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "\n",
    "posG_umap2D = layout_portrait_umap(G,DF_merge_dismod, 2, \n",
    "                                   n_neighbors = n_n, \n",
    "                                   spread = spr, \n",
    "                                   min_dist = md, \n",
    "                                   metric = metr) \n",
    "posG = posG_umap2D"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "angry-morocco",
   "metadata": {
    "tags": []
   },
   "source": [
    "### VISUAL SETTINGS + PLOT PREPARATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "virtual-anchor",
   "metadata": {},
   "source": [
    "#### NODES + EDGES SETTINGS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 621,
   "id": "seven-politics",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "opacity_nodes = 0.8\n",
    "node_edge_col = '#696969' \n",
    "node_linewidth = 1\n",
    "\n",
    "scale_factor = 0.5\n",
    "size_plotly = list(draw_node_degree(G, scale_factor/100).values())\n",
    "\n",
    "color_method = 'DOID:'+num_doid\n",
    "\n",
    "node_col = '#00E3DF' #'#FF8E00' \n",
    "undefined_col = '#DEDEDE'\n",
    "rest_col_nodes = '#DEDEDE'\n",
    "\n",
    "\n",
    "# color nodes\n",
    "col_param = list(df_centralities['closeness'].values)\n",
    "d_node_colors = dict(zip(list(G.nodes()), col_param))\n",
    "\n",
    "c1='#E2E2E2' \n",
    "c2='#4C4C4C' \n",
    "n=len(set(col_param))\n",
    "\n",
    "#fig, ax = plt.subplots(figsize=(8, 5))\n",
    "colors_for_palette = []\n",
    "for x in range(n+1):\n",
    "#    ax.axvline(x, color=colorFader(c1,c2,x/n), linewidth=4)\n",
    "    colors_for_palette.append(colorFader(c1,c2,x/n))    \n",
    "#plt.show()\n",
    "\n",
    "\n",
    "# color edges    \n",
    "edge_color = node_col \n",
    "d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "width_edges = 0.35\n",
    "opacity_edges = 0.35\n",
    "\n",
    "#d_col_all = color_nodes_from_list(G, l_genes, node_col)\n",
    "#colors = list(d_col_all.values())\n",
    "\n",
    "customPalette = sns.set_palette(sns.color_palette(colors_for_palette))\n",
    "d_node_colors = dict(zip(list(G.nodes()), col_param))\n",
    "\n",
    "col_pal = customPalette\n",
    "\n",
    "d_colors = color_nodes_from_dict(G, d_node_colors, palette = col_pal)\n",
    "colors = list(d_colors.values())\n",
    "edge_color = '#d3d3d3'\n",
    "\n",
    "d_col_all = {}\n",
    "for k,v in d_colors.items():\n",
    "    if k in l_genes:\n",
    "        d_col_all[k]=node_col\n",
    "    else:\n",
    "        d_col_all[k]=v\n",
    "        \n",
    "customPalette = sns.set_palette(sns.color_palette(colors_for_palette))\n",
    "d_node_colors = dict(zip(list(G.nodes()), col_param))\n",
    "\n",
    "col_pal = customPalette\n",
    "\n",
    "d_colors = color_nodes_from_dict(G, d_node_colors, palette = col_pal)\n",
    "colors = list(d_colors.values())\n",
    "edge_color = '#d3d3d3'\n",
    "\n",
    "d_col_all = {}\n",
    "for k,v in d_colors.items():\n",
    "    if k in l_genes:\n",
    "        d_col_all[k]=node_col\n",
    "    else:\n",
    "        d_col_all[k]=v"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "twenty-little",
   "metadata": {
    "tags": []
   },
   "source": [
    "### PLOT NETWORK PORTRAIT - plotly "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 622,
   "id": "statistical-protein",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# plot nodes based on focus > separate to background/foreground \n",
    "\n",
    "posG_foreground = {}\n",
    "posG_background = {}\n",
    "for k,v in posG.items():\n",
    "    if k in l_genes:\n",
    "        posG_foreground[k]=v\n",
    "    else: \n",
    "        posG_background[k]=v\n",
    "    \n",
    "d_colors_foreground = {}\n",
    "d_colors_background = {}\n",
    "for i,c in d_col_all.items():\n",
    "    if i in posG_foreground.keys():\n",
    "        d_colors_foreground[i]=c\n",
    "    else: \n",
    "        d_colors_background[i]=c\n",
    "\n",
    "colors_foreground = list(d_colors_foreground.values())\n",
    "colors_background = list(d_colors_background.values())\n",
    "\n",
    "d_feat_foreground = {}\n",
    "d_feat_background = {}\n",
    "for k,v in d_gene_sym.items():\n",
    "    if k in posG_foreground: \n",
    "        d_feat_foreground[k]=v\n",
    "    else:\n",
    "        d_feat_background[k]=v\n",
    "        \n",
    "feat_foreground = list(d_feat_foreground.values())\n",
    "feat_background = list(d_feat_background.values())\n",
    "\n",
    "d_size_plotly = draw_node_degree(G, scale_factor/100)\n",
    "d_size_plotly_foreground = {}\n",
    "d_size_plotly_background = {}\n",
    "for k,v in d_size_plotly.items():\n",
    "    if k in posG_foreground.keys():\n",
    "        d_size_plotly_foreground[k]=v\n",
    "    else:\n",
    "        d_size_plotly_background[k]=v\n",
    "        \n",
    "size_plotly_foreground = list(d_size_plotly_foreground.values())\n",
    "size_plotly_background = list(d_size_plotly_background.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 623,
   "id": "protecting-cloud",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'output_plots/2Dlayouts/2Dportrait_NetlayoutDisease_DOID962_enh15_enh25_nn8_spr1.0_md0.0_cosine_human.html'"
      ]
     },
     "execution_count": 623,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "umap_nodes_foreground = get_trace_nodes_2D(posG_foreground, feat_foreground, colors_foreground, size_plotly_foreground, \n",
    "                                           node_linewidth*0.25, 0.8)\n",
    "umap_nodes_foreground_ = get_trace_nodes_2D(posG_foreground, feat_foreground, colors_foreground, 10,\n",
    "                                           None, 0.2)\n",
    "umap_nodes_background = get_trace_nodes_2D(posG_background, feat_background, colors_background, size_plotly_background, \n",
    "                                           None,0.5) #node_linewidth*0.05, 0.5)\n",
    "umap_nodes_background_ = get_trace_nodes_2D(posG_background, feat_background, colors_background, 2, \n",
    "                                            None,0.5)\n",
    "\n",
    "umap_edges = get_trace_edges_specific2D(d_edge_col, posG, linew=width_edges, opac=opacity_edges)\n",
    "data = [umap_edges, \n",
    "        umap_nodes_background_, \n",
    "        umap_nodes_background, \n",
    "        umap_nodes_foreground_,\n",
    "        umap_nodes_foreground]\n",
    "\n",
    "path = 'output_plots/2Dlayouts/'\n",
    "schema = 'light' #'dark' \n",
    "fname = '2Dportrait_NetlayoutDisease_'+dismod_id+'_enh1'+str(scalar_val_1)+'_enh2'+str(scalar_val_2)+'_nn'+str(n_n)+'_spr'+str(spr)+'_md'+str(md)+'_'+str(metr)+'_'+organism\n",
    "\n",
    "plot_2D(data,path,fname)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "disturbed-saudi",
   "metadata": {},
   "source": [
    "# 3. | NF1 in Disease Context \n",
    "+ other diseases to hightlight e.g. \n",
    "+ DOID:37 - Skin Disease / DOID:3165 - Skin benign neoplasm\n",
    "+ DOID:83 - Cataract (lens disease where eye lens is clouding)\n",
    "+ DOID:863 - Nervous System Disease / DOID:438 - autoimmune disease of the nervous system\n",
    "+ DOID:767 - Muscular Atrophy\n",
    "+ resources: https://disease-ontology.org/ https://en.wikipedia.org/wiki/Neurofibromatosis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 596,
   "id": "nutritional-finding",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENHANCING FACTOR:  0.468785\n"
     ]
    }
   ],
   "source": [
    "# --------------------------------------\n",
    "#\n",
    "scalar_val = 1\n",
    "#\n",
    "# --------------------------------------\n",
    "\n",
    "enhance_factor = round((1-l_max_visprob)*scalar_val_1,10)\n",
    "print('ENHANCING FACTOR: ',enhance_factor)\n",
    "\n",
    "n_n = 8 #6 #20\n",
    "spr = 1.0\n",
    "md = 0.0 #0.4 #0.4\n",
    "metr = 'cosine'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 599,
   "id": "macro-assessment",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get Disease Feature Matrix\n",
    "\n",
    "FM_disease = pd.read_csv('input/Features_Disease_Dataframe_'+organism+'.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 602,
   "id": "sapphire-mounting",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Genes associated to DOID 3165 : 163\n",
      "All genes: 163\n"
     ]
    }
   ],
   "source": [
    "# First Disease Vector \n",
    "# -----------------------------\n",
    "num_doid_1 = '3165'\n",
    "# -----------------------------\n",
    "\n",
    "# define specific DOID \n",
    "dismod_1 = FM_disease.loc[:,'DOID:'+num_doid_1]\n",
    "dismod_1.index = FM_disease.index\n",
    "dismod_id_1 = 'DOID'+num_doid_1\n",
    "\n",
    "dismod_genes_1 = [] \n",
    "for i,v in enumerate(dismod_1.values):\n",
    "    if v == 1:\n",
    "        dismod_genes_1.append(dismod_1.index[i])\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "print('Number of Genes associated to DOID', num_doid_1,':',len(dismod_genes_1))\n",
    "\n",
    "l_genes_1 = [str(i) for i in dismod_genes_1]\n",
    "print('All genes:', len(l_genes_1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 653,
   "id": "proved-floating",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([389289,   4524,   4353,  27127,   5925,  10735,  26259,   1029,\n",
       "              9649,   1630,\n",
       "            ...\n",
       "             65125,   6787,  90678,  10661,  56479,  63910,   7287,  11173,\n",
       "              9962,   5973],\n",
       "           dtype='int64', length=6538)"
      ]
     },
     "execution_count": 653,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dismod_1.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 665,
   "id": "running-abraham",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "389289    0\n",
      "4524      0\n",
      "4353      0\n",
      "27127     0\n",
      "5925      0\n",
      "         ..\n",
      "63910     0\n",
      "7287      0\n",
      "11173     0\n",
      "9962      0\n",
      "5973      0\n",
      "Name: DOID:3165, Length: 6538, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "for i in dismod_1.index:\n",
    "    if i == 3759:\n",
    "        print(dismod_1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "wired-harvard",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Vector for Disease Module 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 666,
   "id": "transsexual-graham",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DisMod 1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66008</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8473</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2561</th>\n",
       "      <td>0.468785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3759</th>\n",
       "      <td>0.468785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22906</th>\n",
       "      <td>0.468785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143903</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10861</th>\n",
       "      <td>0.468785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51471</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221044</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29965</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16376 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        DisMod 1\n",
       "66008   0.000000\n",
       "8473    0.000000\n",
       "2561    0.468785\n",
       "3759    0.468785\n",
       "22906   0.468785\n",
       "...          ...\n",
       "143903  0.000000\n",
       "10861   0.468785\n",
       "51471   0.000000\n",
       "221044  0.000000\n",
       "29965   0.000000\n",
       "\n",
       "[16376 rows x 1 columns]"
      ]
     },
     "execution_count": 666,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create an empty matrix for zeros of rest genes (not associated to any disease)\n",
    "rest = []\n",
    "for i in G.nodes():\n",
    "    if str(i) in dismod_1.index or int(i) in dismod_1.index:\n",
    "        pass\n",
    "    else: \n",
    "        rest.append(i)\n",
    "        \n",
    "empty = np.zeros(shape=(len(rest),1))\n",
    "empty_stacked = np.vstack(empty)\n",
    "df_empty = pd.DataFrame(empty_stacked)\n",
    "df_empty.index = rest\n",
    "\n",
    "# combine disease vector with rest of genes in graph\n",
    "df_dismod_complete_1 = pd.concat([dismod_1, df_empty])\n",
    "\n",
    "# reorder according to G.nodes\n",
    "df_dismod_complete_1.columns = ['DisMod 1']\n",
    "DF_dismod_reorder_1 = df_dismod_complete_1.reindex(list(G.nodes()))\n",
    "\n",
    "# fill nan positions (due to reordering) with 1 \n",
    "DF_dismod_new_1 = DF_dismod_reorder_1.replace(np.nan,1)\n",
    "DF_dismod_scal_1 = DF_dismod_new_1 * enhance_factor\n",
    "DF_dismod_scal_1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 603,
   "id": "human-fighter",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Genes associated to DOID 83 : 277\n",
      "All genes: 277\n"
     ]
    }
   ],
   "source": [
    "# Second Disease Vector \n",
    "# -----------------------------\n",
    "num_doid_2 = '83'\n",
    "# -----------------------------\n",
    "\n",
    "# define specific DOID \n",
    "dismod_2 = FM_disease.loc[:,'DOID:'+num_doid_2]\n",
    "dismod_2.index = FM_disease.index\n",
    "dismod_id_2 = 'DOID'+num_doid_2\n",
    "\n",
    "dismod_genes_2 = [] \n",
    "for i,v in enumerate(dismod_2.values):\n",
    "    if v == 1:\n",
    "        dismod_genes_2.append(dismod_2.index[i])\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "print('Number of Genes associated to DOID', num_doid_2,':',len(dismod_genes_2))\n",
    "\n",
    "l_genes_2 = [str(i) for i in dismod_genes_2]\n",
    "print('All genes:', len(l_genes_2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "characteristic-yugoslavia",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "familiar-auckland",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Vector for Disease Module 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 635,
   "id": "personal-internship",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease Module</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66008</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8473</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2561</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3759</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22906</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143903</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10861</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51471</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221044</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29965</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16376 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Disease Module\n",
       "66008              0.0\n",
       "8473               0.0\n",
       "2561               0.0\n",
       "3759               0.0\n",
       "22906              0.0\n",
       "...                ...\n",
       "143903             0.0\n",
       "10861              0.0\n",
       "51471              0.0\n",
       "221044             0.0\n",
       "29965              0.0\n",
       "\n",
       "[16376 rows x 1 columns]"
      ]
     },
     "execution_count": 635,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create an empty matrix for zeros of rest genes (not associated to any disease)\n",
    "rest = []\n",
    "for i in G.nodes():\n",
    "    if str(i) in dismod_genes_2 or int(i) in dismod_genes_2:\n",
    "        pass\n",
    "    else: \n",
    "        rest.append(i)\n",
    "        \n",
    "empty = np.zeros(shape=(len(rest),1))\n",
    "empty_stacked = np.vstack(empty)\n",
    "df_empty = pd.DataFrame(empty_stacked)\n",
    "df_empty.index = rest\n",
    "\n",
    "# combine disease vector with rest of genes in graph\n",
    "df_dismod_complete_2 = pd.concat([dismod_2, df_empty])\n",
    "\n",
    "# reorder according to G.nodes\n",
    "df_dismod_complete_2.columns = ['DisMod 2']\n",
    "DF_dismod_reorder_2 = df_dismod_complete_2.reindex(list(G.nodes()))\n",
    "\n",
    "# fill nan positions (due to reordering) with 1 \n",
    "DF_dismod_new_2 = DF_dismod_reorder_2.replace(np.nan,1)\n",
    "DF_dismod_scal_2 = DF_dismod_new_2 * enhance_factor\n",
    "DF_dismod_scal_2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "disabled-pennsylvania",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 605,
   "id": "built-slovakia",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Genes associated to DOID 438 : 8\n",
      "All genes: 8\n"
     ]
    }
   ],
   "source": [
    "# Third Disease Vector \n",
    "# -----------------------------\n",
    "num_doid_3 = '438'\n",
    "# -----------------------------\n",
    "\n",
    "# define specific DOID \n",
    "dismod_3 = FM_disease.loc[:,'DOID:'+num_doid_3]\n",
    "dismod_3.index = FM_disease.index\n",
    "dismod_id_3 = 'DOID'+num_doid_3\n",
    "\n",
    "dismod_genes_3 = [] \n",
    "for i,v in enumerate(dismod_3.values):\n",
    "    if v == 1:\n",
    "        dismod_genes_3.append(dismod_3.index[i])\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "print('Number of Genes associated to DOID', num_doid_3,':',len(dismod_genes_3))\n",
    "\n",
    "l_genes_3 = [str(i) for i in dismod_genes_3]\n",
    "print('All genes:', len(l_genes_3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "naked-rochester",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "right-first",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "precise-works",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "violent-perth",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "voluntary-logging",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "improved-cherry",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "framed-carpet",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 607,
   "id": "loaded-sociology",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Genes associated to DOID 767 : 149\n",
      "All genes: 149\n"
     ]
    }
   ],
   "source": [
    "# Fourth Disease Vector \n",
    "# -----------------------------\n",
    "num_doid_4 = '767'\n",
    "# -----------------------------\n",
    "\n",
    "# define specific DOID \n",
    "dismod_4 = FM_disease.loc[:,'DOID:'+num_doid_4]\n",
    "dismod_4.index = FM_disease.index\n",
    "dismod_id_4 = 'DOID'+num_doid_4\n",
    "\n",
    "dismod_genes_4 = [] \n",
    "for i,v in enumerate(dismod_4.values):\n",
    "    if v == 1:\n",
    "        dismod_genes_4.append(dismod_4.index[i])\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "print('Number of Genes associated to DOID', num_doid_4,':',len(dismod_genes_4))\n",
    "\n",
    "l_genes_4 = [str(i) for i in dismod_genes_4]\n",
    "print('All genes:', len(l_genes_4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 626,
   "id": "amazing-virtue",
   "metadata": {},
   "outputs": [],
   "source": [
    "dismod_all = [dismod_genes_1,dismod_genes_2 ] #,dismod_genes_3,dismod_genes_4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 632,
   "id": "sunrise-measurement",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1029, 7015, 2033, 207, 4893, 8085, 6648, 3845, 3479, 3265, 1499, 472, 5290, 1440, 2068, 5743, 2261, 8314, 2175, 360, 7157, 1387, 7403, 5728, 5970, 6392, 4591, 5894, 673, 5573, 3481, 5979, 7249, 7248, 4000, 2321, 7428, 4286, 2952, 6390, 2736, 2263, 6667, 3552, 4254, 11200, 4851, 1032, 6608, 4878, 4693, 1460, 2052, 6391, 7046, 5781, 7507, 1030, 3713, 1457, 5395, 5921, 434, 4763, 2697, 5605, 5159, 2185, 7517, 3480, 2776, 2066, 8626, 2324, 2178, 3872, 2993, 7474, 940, 2176, 5604, 1459, 3791, 51684, 2706, 4292, 2709, 4854, 4780, 4089, 4318, 841, 4436, 1540, 3858, 51176, 5155, 4193, 4255, 7299, 4683, 79784, 5467, 4953, 2248, 867, 23095, 2177, 10654, 7040, 54386, 4952, 1277, 7133, 5727, 843, 9968, 641, 3091, 2658, 1050, 326, 4221, 4913, 1019, 8643, 3848, 4157, 4014, 23512, 2956, 4507, 4352, 10483, 1385, 1493, 25913, 4595, 7010, 1289, 2271, 50485, 4534, 6663, 23592, 2022, 3257, 55131, 2717, 94, 5017, 5333, 1855, 5116, 55120, 2517, 4920, 58495, 25942, 1281, 8398, 4126, 10484]\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "cannot concatenate object of type '<class 'list'>'; only Series and DataFrame objs are valid",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-632-402689c5a8e6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m     \u001b[0;31m# combine disease vector with rest of genes in graph\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m     \u001b[0mdf_dismod_complete\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdismod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf_empty\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m     \u001b[0;31m# reorder according to G.nodes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/myenv_clip2/lib/python3.6/site-packages/pandas/core/reshape/concat.py\u001b[0m in \u001b[0;36mconcat\u001b[0;34m(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)\u001b[0m\n\u001b[1;32m    279\u001b[0m         \u001b[0mverify_integrity\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mverify_integrity\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    280\u001b[0m         \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 281\u001b[0;31m         \u001b[0msort\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    282\u001b[0m     )\n\u001b[1;32m    283\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/myenv_clip2/lib/python3.6/site-packages/pandas/core/reshape/concat.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, objs, axis, join, keys, levels, names, ignore_index, verify_integrity, copy, sort)\u001b[0m\n\u001b[1;32m    355\u001b[0m                     \u001b[0;34m\"only Series and DataFrame objs are valid\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtyp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    356\u001b[0m                 )\n\u001b[0;32m--> 357\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    358\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    359\u001b[0m             \u001b[0;31m# consolidate\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: cannot concatenate object of type '<class 'list'>'; only Series and DataFrame objs are valid"
     ]
    }
   ],
   "source": [
    "DF_all = []\n",
    "\n",
    "for i,dismod in enumerate(dismod_all):\n",
    "\n",
    "    # create an empty matrix for zeros of rest genes (not associated to any disease)\n",
    "    rest = []\n",
    "    for i in G.nodes():\n",
    "        if i in dismod or int(i) in dismod:\n",
    "            pass\n",
    "        else: \n",
    "            rest.append(i)\n",
    "    empty = np.zeros(shape=(len(rest),1))\n",
    "    empty_stacked = np.vstack(empty)\n",
    "    df_empty = pd.DataFrame(empty_stacked)\n",
    "    df_empty.index = rest\n",
    "    \n",
    "    # combine disease vector with rest of genes in graph\n",
    "    df_dismod_complete = pd.concat([dismod, df_empty])\n",
    "\n",
    "    # reorder according to G.nodes\n",
    "    df_dismod_complete.columns = ['First']\n",
    "    DF_dismod_reorder = df_dismod_complete.reindex(list(G.nodes()))\n",
    "\n",
    "    # fill nan positions (due to reordering) with 1 \n",
    "    DF_dismod_new = DF_dismod_reorder.replace(np.nan,1)\n",
    "\n",
    "    d_vec_disease = {}\n",
    "    for i in G.nodes():\n",
    "        if i in dismod:\n",
    "            d_vec_disease[i]=1\n",
    "        else:\n",
    "            d_vec_disease[i]=0\n",
    "\n",
    "    d_vec_disease_sorted = {key:d_vec_disease[key] for key in G.nodes()}\n",
    "    DF = pd.DataFrame(list(d_vec_disease_sorted.values()))\n",
    "    DF.index = list(G.nodes())\n",
    "    DF.columns = [str(i)]\n",
    "\n",
    "    DF_all.append(DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "surprised-resident",
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_dismod_all = [] \n",
    "\n",
    "for i in DF_all: \n",
    "    DF = i * enhance_factor\n",
    "    DF_dismod_all.append(DF)\n",
    "DF_merge_dismod = pd.concat([DF_dismod_scal],axis=1)\n",
    "\n",
    "DF_merge_dismod.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "unable-lightweight",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faced-editor",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bottom-member",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "psychological-basis",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "rocky-calcium",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "considerable-panel",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "affected-ballot",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "unlikely-custom",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "brave-iraqi",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "posted-optimum",
   "metadata": {},
   "source": [
    "# 4. | STRUC + FUNC (complete disease matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "id": "worldwide-clock",
   "metadata": {},
   "outputs": [],
   "source": [
    "FM_disease = pd.read_csv('input/Features_Disease_Dataframe_'+organism+'.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "id": "practical-heating",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENHANCING FACTOR functional features:  0.375028\n"
     ]
    }
   ],
   "source": [
    "# --------------------------------------\n",
    "#\n",
    "scalar_val = 4\n",
    "#\n",
    "# --------------------------------------\n",
    "\n",
    "enhance_factor = round((1-l_max_visprob)*scalar_val,10) # for emphasizing functional features > should be >= max of structural matrix values  # if significantly higher than max values in structural matrix > causes isolation of nodes with functional features enhanced\n",
    "print('ENHANCING FACTOR functional features: ',enhance_factor)\n",
    "\n",
    "n_n = 14\n",
    "spr = 1\n",
    "md = 0.0\n",
    "metr = 'cosine'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "id": "demographic-foster",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create an empty matrix for zeros of rest genes (not associated to any disease)\n",
    "rest = []\n",
    "for i in G.nodes():\n",
    "    if str(i) in FM_disease.index or int(i) in FM_disease.index:\n",
    "        pass\n",
    "    else: \n",
    "        rest.append(str(i))\n",
    "        \n",
    "empty = np.zeros(shape=(len(rest),len(FM_disease.columns)))\n",
    "empty_stacked = np.vstack(empty)\n",
    "df_empty = pd.DataFrame(empty_stacked)\n",
    "df_empty.index = rest\n",
    "df_empty.columns = FM_disease.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "id": "sustainable-photographer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DOID:11054</th>\n",
       "      <th>DOID:3996</th>\n",
       "      <th>DOID:0050686</th>\n",
       "      <th>DOID:162</th>\n",
       "      <th>DOID:14566</th>\n",
       "      <th>DOID:4</th>\n",
       "      <th>DOID:615</th>\n",
       "      <th>DOID:9500</th>\n",
       "      <th>DOID:74</th>\n",
       "      <th>DOID:7</th>\n",
       "      <th>...</th>\n",
       "      <th>DOID:0060485</th>\n",
       "      <th>DOID:12577</th>\n",
       "      <th>DOID:732</th>\n",
       "      <th>DOID:0050816</th>\n",
       "      <th>DOID:0060589</th>\n",
       "      <th>DOID:0060202</th>\n",
       "      <th>DOID:14116</th>\n",
       "      <th>DOID:3137</th>\n",
       "      <th>DOID:100</th>\n",
       "      <th>DOID:0110648</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>389289</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4524</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4353</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27127</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5925</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51458</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143903</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51471</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221044</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29965</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16376 rows × 3911 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        DOID:11054  DOID:3996  DOID:0050686  DOID:162  DOID:14566  DOID:4  \\\n",
       "389289         1.0        1.0           1.0       1.0         1.0     1.0   \n",
       "4524           1.0        1.0           1.0       1.0         1.0     1.0   \n",
       "4353           1.0        1.0           1.0       1.0         1.0     1.0   \n",
       "27127          1.0        1.0           1.0       1.0         1.0     1.0   \n",
       "5925           1.0        1.0           1.0       1.0         1.0     1.0   \n",
       "...            ...        ...           ...       ...         ...     ...   \n",
       "51458          0.0        0.0           0.0       0.0         0.0     0.0   \n",
       "143903         0.0        0.0           0.0       0.0         0.0     0.0   \n",
       "51471          0.0        0.0           0.0       0.0         0.0     0.0   \n",
       "221044         0.0        0.0           0.0       0.0         0.0     0.0   \n",
       "29965          0.0        0.0           0.0       0.0         0.0     0.0   \n",
       "\n",
       "        DOID:615  DOID:9500  DOID:74  DOID:7  ...  DOID:0060485  DOID:12577  \\\n",
       "389289       0.0        0.0      0.0     0.0  ...           0.0         0.0   \n",
       "4524         1.0        1.0      1.0     1.0  ...           0.0         0.0   \n",
       "4353         0.0        0.0      1.0     1.0  ...           0.0         0.0   \n",
       "27127        0.0        0.0      0.0     0.0  ...           0.0         0.0   \n",
       "5925         0.0        0.0      0.0     1.0  ...           0.0         0.0   \n",
       "...          ...        ...      ...     ...  ...           ...         ...   \n",
       "51458        0.0        0.0      0.0     0.0  ...           0.0         0.0   \n",
       "143903       0.0        0.0      0.0     0.0  ...           0.0         0.0   \n",
       "51471        0.0        0.0      0.0     0.0  ...           0.0         0.0   \n",
       "221044       0.0        0.0      0.0     0.0  ...           0.0         0.0   \n",
       "29965        0.0        0.0      0.0     0.0  ...           0.0         0.0   \n",
       "\n",
       "        DOID:732  DOID:0050816  DOID:0060589  DOID:0060202  DOID:14116  \\\n",
       "389289       0.0           0.0           0.0           0.0         0.0   \n",
       "4524         0.0           0.0           0.0           0.0         0.0   \n",
       "4353         0.0           0.0           0.0           0.0         0.0   \n",
       "27127        0.0           0.0           0.0           0.0         0.0   \n",
       "5925         0.0           0.0           0.0           0.0         0.0   \n",
       "...          ...           ...           ...           ...         ...   \n",
       "51458        0.0           0.0           0.0           0.0         0.0   \n",
       "143903       0.0           0.0           0.0           0.0         0.0   \n",
       "51471        0.0           0.0           0.0           0.0         0.0   \n",
       "221044       0.0           0.0           0.0           0.0         0.0   \n",
       "29965        0.0           0.0           0.0           0.0         0.0   \n",
       "\n",
       "        DOID:3137  DOID:100  DOID:0110648  \n",
       "389289        0.0       0.0           0.0  \n",
       "4524          0.0       0.0           0.0  \n",
       "4353          0.0       0.0           0.0  \n",
       "27127         0.0       0.0           0.0  \n",
       "5925          0.0       0.0           0.0  \n",
       "...           ...       ...           ...  \n",
       "51458         0.0       0.0           0.0  \n",
       "143903        0.0       0.0           0.0  \n",
       "51471         0.0       0.0           0.0  \n",
       "221044        0.0       0.0           0.0  \n",
       "29965         0.0       0.0           0.0  \n",
       "\n",
       "[16376 rows x 3911 columns]"
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine disease vector with rest of genes in graph\n",
    "df_diseases_complete = pd.concat([FM_disease, df_empty])\n",
    "df_diseases_complete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "id": "pacific-approach",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>66008</th>\n",
       "      <th>8473</th>\n",
       "      <th>2561</th>\n",
       "      <th>3759</th>\n",
       "      <th>22906</th>\n",
       "      <th>4928</th>\n",
       "      <th>1994</th>\n",
       "      <th>8481</th>\n",
       "      <th>81610</th>\n",
       "      <th>51361</th>\n",
       "      <th>...</th>\n",
       "      <th>DOID:0060485</th>\n",
       "      <th>DOID:12577</th>\n",
       "      <th>DOID:732</th>\n",
       "      <th>DOID:0050816</th>\n",
       "      <th>DOID:0060589</th>\n",
       "      <th>DOID:0060202</th>\n",
       "      <th>DOID:14116</th>\n",
       "      <th>DOID:3137</th>\n",
       "      <th>DOID:100</th>\n",
       "      <th>DOID:0110648</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66008</th>\n",
       "      <td>0.900242</td>\n",
       "      <td>0.002519</td>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.002525</td>\n",
       "      <td>0.002517</td>\n",
       "      <td>0.002512</td>\n",
       "      <td>0.002511</td>\n",
       "      <td>0.002502</td>\n",
       "      <td>0.002501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8473</th>\n",
       "      <td>0.000687</td>\n",
       "      <td>0.900245</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2561</th>\n",
       "      <td>0.012865</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.900367</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>...</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3759</th>\n",
       "      <td>0.006433</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.900389</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>...</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22906</th>\n",
       "      <td>0.005347</td>\n",
       "      <td>0.005316</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.900120</td>\n",
       "      <td>0.005314</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143903</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10861</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "      <td>0.375028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51471</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221044</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29965</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16376 rows × 20287 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           66008      8473      2561      3759     22906      4928      1994  \\\n",
       "66008   0.900242  0.002519  0.002502  0.002502  0.002525  0.002517  0.002512   \n",
       "8473    0.000687  0.900245  0.000002  0.000002  0.000685  0.000008  0.000031   \n",
       "2561    0.012865  0.000036  0.900367  0.000036  0.000036  0.000065  0.000037   \n",
       "3759    0.006433  0.000018  0.000018  0.900389  0.000018  0.000027  0.000041   \n",
       "22906   0.005347  0.005316  0.000015  0.000015  0.900120  0.005314  0.000026   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "143903  0.000000  0.000000  0.000001  0.000000  0.000000  0.000000  0.000004   \n",
       "10861   0.000000  0.000000  0.000001  0.000000  0.000000  0.000000  0.000004   \n",
       "51471   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000054   \n",
       "221044  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000003   \n",
       "29965   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000002   \n",
       "\n",
       "            8481     81610     51361  ...  DOID:0060485  DOID:12577  DOID:732  \\\n",
       "66008   0.002511  0.002502  0.002501  ...      0.000000    0.000000  0.000000   \n",
       "8473    0.000006  0.000002  0.000002  ...      0.000000    0.000000  0.000000   \n",
       "2561    0.000036  0.000036  0.000036  ...      0.375028    0.375028  0.375028   \n",
       "3759    0.000018  0.000018  0.000018  ...      0.375028    0.375028  0.375028   \n",
       "22906   0.000015  0.000015  0.000016  ...      0.375028    0.375028  0.375028   \n",
       "...          ...       ...       ...  ...           ...         ...       ...   \n",
       "143903  0.000000  0.000000  0.000000  ...      0.000000    0.000000  0.000000   \n",
       "10861   0.000000  0.000000  0.000000  ...      0.375028    0.375028  0.375028   \n",
       "51471   0.000000  0.000000  0.000000  ...      0.000000    0.000000  0.000000   \n",
       "221044  0.000000  0.000000  0.000000  ...      0.000000    0.000000  0.000000   \n",
       "29965   0.000046  0.000000  0.000000  ...      0.000000    0.000000  0.000000   \n",
       "\n",
       "        DOID:0050816  DOID:0060589  DOID:0060202  DOID:14116  DOID:3137  \\\n",
       "66008       0.000000      0.000000      0.000000    0.000000   0.000000   \n",
       "8473        0.000000      0.000000      0.000000    0.000000   0.000000   \n",
       "2561        0.375028      0.375028      0.375028    0.375028   0.375028   \n",
       "3759        0.375028      0.375028      0.375028    0.375028   0.375028   \n",
       "22906       0.375028      0.375028      0.375028    0.375028   0.375028   \n",
       "...              ...           ...           ...         ...        ...   \n",
       "143903      0.000000      0.000000      0.000000    0.000000   0.000000   \n",
       "10861       0.375028      0.375028      0.375028    0.375028   0.375028   \n",
       "51471       0.000000      0.000000      0.000000    0.000000   0.000000   \n",
       "221044      0.000000      0.000000      0.000000    0.000000   0.000000   \n",
       "29965       0.000000      0.000000      0.000000    0.000000   0.000000   \n",
       "\n",
       "        DOID:100  DOID:0110648  \n",
       "66008   0.000000      0.000000  \n",
       "8473    0.000000      0.000000  \n",
       "2561    0.375028      0.375028  \n",
       "3759    0.375028      0.375028  \n",
       "22906   0.375028      0.375028  \n",
       "...          ...           ...  \n",
       "143903  0.000000      0.000000  \n",
       "10861   0.375028      0.375028  \n",
       "51471   0.000000      0.000000  \n",
       "221044  0.000000      0.000000  \n",
       "29965   0.000000      0.000000  \n",
       "\n",
       "[16376 rows x 20287 columns]"
      ]
     },
     "execution_count": 334,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reorder according to G.nodes\n",
    "DF_diseases_reorder = df_diseases_complete.reindex(DF_structural.index)\n",
    "\n",
    "# fill nan positions (due to reordering) with 1 \n",
    "DF_diseases_new = DF_diseases_reorder.replace(np.nan,1)\n",
    "\n",
    "# multiply with enhancing factor\n",
    "DF_diseases_scal = DF_diseases_new*enhance_factor \n",
    "\n",
    "DF_merge_diseases = pd.concat([DF_structural.T, \n",
    "                             DF_diseases_scal],axis=1)\n",
    "\n",
    "DF_merge_diseases"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "tribal-quantity",
   "metadata": {},
   "source": [
    "### 2D PORTRAIT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "id": "loving-placement",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_n = 12\n",
    "spr = 1\n",
    "md = 0.0\n",
    "metr = 'cosine'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "id": "falling-stereo",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min 9s, sys: 15.3 s, total: 3min 25s\n",
      "Wall time: 1min 8s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "\n",
    "posG_umap2D = layout_portrait_umap(G,DF_merge_diseases, 2,\n",
    "                                   n_neighbors = n_n,\n",
    "                                   spread = spr,\n",
    "                                   min_dist = md,\n",
    "                                   metric = metr) \n",
    "posG = posG_umap2D"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "welsh-marina",
   "metadata": {
    "tags": []
   },
   "source": [
    "## VISUAL SETTINGS + PLOT PREPARATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "intellectual-ensemble",
   "metadata": {},
   "source": [
    "#### NODES - GENERAL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "id": "removable-matthew",
   "metadata": {},
   "outputs": [],
   "source": [
    "opacity_nodes = 0.8\n",
    "node_edge_col = '#696969' \n",
    "node_linewidth = 1\n",
    "\n",
    "scale_factor = 0.5\n",
    "size_plotly = list(draw_node_degree(G, scale_factor/100).values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "id": "extra-investing",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "color_method = 'NF1'\n",
    "l_genes = enhance_genes\n",
    "\n",
    "node_col = '#00E3DF' #'#FF8E00' \n",
    "undefined_col = '#DEDEDE'\n",
    "rest_col_nodes = '#DEDEDE'\n",
    "\n",
    "d_col_all = color_nodes_from_list(G, l_genes, node_col)\n",
    "colors = list(d_col_all.values())\n",
    "    \n",
    "edge_color = node_col \n",
    "d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "width_edges = 0.25\n",
    "opacity_edges = 0.5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "particular-handling",
   "metadata": {
    "tags": []
   },
   "source": [
    "## PLOT NETWORK PORTRAIT - plotly "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "smoking-renaissance",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# plot nodes based on focus > separate to background/foreground \n",
    "\n",
    "posG_foreground = {}\n",
    "posG_background = {}\n",
    "for k,v in posG.items():\n",
    "    if k in l_genes:\n",
    "        posG_foreground[k]=v\n",
    "    else: \n",
    "        posG_background[k]=v\n",
    "    \n",
    "d_colors_foreground = {}\n",
    "d_colors_background = {}\n",
    "for i,c in d_col_all.items():\n",
    "    if i in posG_foreground.keys():\n",
    "        d_colors_foreground[i]=c\n",
    "    else: \n",
    "        d_colors_background[i]=c\n",
    "\n",
    "colors_foreground = list(d_colors_foreground.values())\n",
    "colors_background = list(d_colors_background.values())\n",
    "\n",
    "d_feat_foreground = {}\n",
    "d_feat_background = {}\n",
    "for k,v in d_gene_sym.items():\n",
    "    if k in posG_foreground: \n",
    "        d_feat_foreground[k]=v\n",
    "    else:\n",
    "        d_feat_background[k]=v\n",
    "        \n",
    "feat_foreground = list(d_feat_foreground.values())\n",
    "feat_background = list(d_feat_background.values())\n",
    "\n",
    "d_size_plotly = draw_node_degree(G, scale_factor/100)\n",
    "d_size_plotly_foreground = {}\n",
    "d_size_plotly_background = {}\n",
    "for k,v in d_size_plotly.items():\n",
    "    if k in posG_foreground.keys():\n",
    "        d_size_plotly_foreground[k]=v\n",
    "    else:\n",
    "        d_size_plotly_background[k]=v\n",
    "        \n",
    "size_plotly_foreground = list(d_size_plotly_foreground.values())\n",
    "size_plotly_background = list(d_size_plotly_background.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "id": "stopped-slovakia",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'output_plots/2Dlayouts/2Dportrait_NetlayoutDisease_DiseaseFeatureMatrix_enhance0.375028_NF1_nn12_spr1_md0.0_cosine_human.html'"
      ]
     },
     "execution_count": 340,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "umap_nodes_foreground = get_trace_nodes_2D(posG_foreground, feat_foreground, colors_foreground, size_plotly_foreground, node_linewidth)\n",
    "umap_nodes_background = get_trace_nodes_2D(posG_background, feat_background, colors_background, size_plotly_background, node_linewidth*0.25)\n",
    "umap_nodes_background_ = get_trace_nodes_2D(posG_background, feat_background, colors_background, 0.5, node_linewidth*0.5)\n",
    "\n",
    "umap_edges = get_trace_edges_specific2D(d_edge_col, posG, linew=width_edges, opac=opacity_edges)\n",
    "data = [umap_edges, umap_nodes_background_, umap_nodes_background, umap_nodes_foreground]\n",
    "\n",
    "path = 'output_plots/2Dlayouts/'\n",
    "schema = 'light' #'dark' \n",
    "fname = '2Dportrait_NetlayoutDisease_DiseaseFeatureMatrix_enhance'+str(enhance_factor)+'_'+color_method+'_nn'+str(n_n)+'_spr'+str(spr)+'_md'+str(md)+'_'+str(metr)+'_'+organism\n",
    "\n",
    "plot_2D(data,path,fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "brutal-extra",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "perceived-lesson",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "heard-colon",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "plain-domain",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "nearby-ownership",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lined-chick",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dedicated-killer",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "corresponding-thickness",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "caring-worthy",
   "metadata": {
    "tags": []
   },
   "source": [
    "### TOPOGRAPHIC MAP "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "synthetic-neighbor",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------------------------------------------------------------------------\n",
    "#\n",
    "# Choose a z-parameter e.g. essentiality, centravg, disease \n",
    "# or choose any dictionary with nodeID: z-value\n",
    "# by setting d_z = {nodeID: val, ...}\n",
    "#\n",
    "# ---------------------------------------------------------------------------\n",
    "z_feat = 'disease'\n",
    "# ---------------------------------------------------------------------------\n",
    "\n",
    "if z_feat == 'essentiality':\n",
    "\n",
    "    ''' Essentiality state of each node displayed on layers of \"essential\", \"non-essential\", \"non-categorized\" ''' \n",
    "\n",
    "    value_ess = 5\n",
    "    value_noness = 3 \n",
    "    value_notdef = 1\n",
    "\n",
    "    d_ess = {}\n",
    "    for i in essential_genes:\n",
    "        d_ess[i] = value_ess\n",
    "    d_noness = {}\n",
    "    for i in non_ess_genes:\n",
    "        d_noness[i] = value_noness\n",
    "    d_notdef = {}\n",
    "    for i in notdefined_genes:\n",
    "        d_notdef[i] = value_notdef\n",
    "\n",
    "    d_alless_unsort = {**d_ess,**d_noness,**d_notdef}\n",
    "    d_z = {key:d_alless_unsort[key] for key in G.nodes()}\n",
    "    \n",
    "    l_genes = essential_genes\n",
    "    \n",
    "    node_col = '#00abff' \n",
    "    undefined_col = '#d3d3d3'\n",
    "    rest_col_nodes = '#d3d3d3'\n",
    "\n",
    "    d_col_all = color_nodes_from_list(G, essential_genes, node_col)\n",
    "    colors = list(d_col_all.values())\n",
    "    \n",
    "    edge_color = '#ACACAC' \n",
    "    d_edge_col = color_edges_from_nodelist_specific(G, l_genes, node_col)\n",
    "    width_edges = 0.15\n",
    "    opacity_edges = 0.1\n",
    "    \n",
    "    umap_nodes = get_trace_nodes_3D(posG, l_features, colors, size, opacity_nodes)\n",
    "    umap_edges = get_trace_edges_specific3D(d_edge_col, posG, linew=width_edges, opac=opacity_edges)  \n",
    "    data = [umap_edges, umap_nodes]\n",
    "    \n",
    "    \n",
    "elif z_feat == 'centravg':\n",
    "    \n",
    "    ''' Average of Centrality Metrics displayed in z-height ''' \n",
    "    \n",
    "    d_clos = {}\n",
    "    for k,v in d_centralities.items():\n",
    "        d_clos[k] = v[1]\n",
    "\n",
    "    d_nodecolors = d_clos \n",
    "    col_pal = 'YlOrRd'\n",
    "\n",
    "    d_colors = color_nodes_from_dict(G, d_nodecolors, palette = col_pal)\n",
    "    colors = list(d_colors.values())\n",
    "    \n",
    "    edge_color = '#ACACAC' \n",
    "    width_edges = 0.15\n",
    "    opacity_edges = 0.1\n",
    "    \n",
    "    umap_nodes = get_trace_nodes_3D(posG, l_features, colors, size, opacity_nodes)\n",
    "    umap_edges = get_trace_edges_3D(G, posG, edge_color, opac = 0.05, linewidth = 0.5)\n",
    "    data = [umap_edges, umap_nodes]\n",
    "    \n",
    "    \n",
    "elif z_feat == 'disease':\n",
    "    \n",
    "    ''' Disease count associated to a node displayed in z-height ''' \n",
    "\n",
    "    DM_feature  = pd.read_csv('input/Features_Disease_Dataframe_Human.csv', index_col=0)\n",
    "\n",
    "    # set gene list (= G.nodes())\n",
    "    genes_assoc = []\n",
    "    for i in DM_feature.index:\n",
    "        genes_assoc.append(str(i))\n",
    "\n",
    "    genes_notassoc = [] \n",
    "    for g in G.nodes():\n",
    "        if g not in genes_assoc:\n",
    "            genes_notassoc.append(g)\n",
    "\n",
    "    features_counted = (DM_feature == 1).astype(int).sum(axis=1)\n",
    "    d_features_counted = dict(zip(genes_assoc, list(features_counted)))\n",
    "    \n",
    "    d_rest = {}\n",
    "    for i in genes_notassoc: \n",
    "        d_rest[i] = -20\n",
    "\n",
    "    d_param_unsorted = {**d_features_counted, **d_rest}\n",
    "    d_z = {key:d_param_unsorted[key] for key in G.nodes()}\n",
    "    \n",
    "    d_clos = {}\n",
    "    for k,v in d_centralities.items():\n",
    "        d_clos[k] = v[1]\n",
    "\n",
    "    d_nodecolors = d_clos \n",
    "    col_pal = 'YlOrRd'\n",
    "\n",
    "    d_colors = color_nodes_from_dict(G, d_nodecolors, palette = col_pal)\n",
    "    colors = list(d_colors.values())\n",
    "    edge_color = '#ACACAC'\n",
    "    width_edges = 0.15\n",
    "    opacity_edges = 0.1\n",
    "\n",
    "    umap_nodes = get_trace_nodes_3D(posG, l_features, colors, size, opacity_nodes)\n",
    "    umap_edges = get_trace_edges_3D(G, posG, edge_color, opac = 0.05, linewidth = 0.5)\n",
    "    data = [umap_edges, umap_nodes]\n",
    "    \n",
    "else:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bigger-miniature",
   "metadata": {},
   "source": [
    "#### PLOT "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "northern-infrastructure",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'output_plots/Topographic/topographic_z_disease_local_human_light.html'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ---------------------------------------------------------------------------\n",
    "#\n",
    "# Choose a 2D layout as basis\n",
    "#\n",
    "# 1. either calculate one e.g.: \n",
    "# posG_2Dportrait = layout_portrait_umap(G,DM,2,n_neighbors=20, spread=1, min_dist=0.0, metric='cosine') \n",
    "# ---------------------------------------------------------------------------\n",
    "#\n",
    "# 2. or use a precalculated 2D layout \n",
    "posG_2D = posG_umap2D\n",
    "# ---------------------------------------------------------------------------\n",
    "\n",
    "posG = layout_topographic(posG_2D, d_z)\n",
    "\n",
    "path = 'output_plots/Topographic/'\n",
    "schema = 'light' #'dark'\n",
    "fname = 'topographic_z_'+z_feat+'_'+netlayout+'_'+organism+'_'+schema\n",
    "\n",
    "plot_3D(data,path,fname, schema)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "front-positive",
   "metadata": {},
   "source": [
    "### GEODESIC MAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "final-chile",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Spherical Shells: 4\n"
     ]
    }
   ],
   "source": [
    "# ---------------------------------------------------------------------------\n",
    "#\n",
    "# Choose a radius-parameter e.g. autocore, essentiality\n",
    "# or choose any dictionary with nodeID: z-value\n",
    "# by setting d_radius = {nodeID: val, ...}\n",
    "#\n",
    "# ---------------------------------------------------------------------------\n",
    "r_feat = 'autocore'\n",
    "# ---------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "if r_feat == 'autocore':\n",
    "\n",
    "    # Load Gene Information for Gene lists \n",
    "    variants = pd.read_csv('input/julias_variants_mod.txt')\n",
    "    variant_list = [str(i) for i in list(variants['variants'])]\n",
    "    variant_sym = pd.read_csv('input/julias_variants_sym.txt')\n",
    "    genes_sym_list = [str(i) for i in list(variant_sym['variant_sym'])]\n",
    "    d_gene_sym_variants = dict(zip(variant_list,genes_sym_list))\n",
    "\n",
    "    df_seeds = pd.read_csv('input/seeds_from_genelists.txt')\n",
    "    df_seeds.columns = ['seeds']\n",
    "    df_seeds_z = pd.read_csv('input/seeds_from_zNumbers.txt', sep='\\n')\n",
    "    seeds_list = [str(i) for i in list(df_seeds['seeds'])]\n",
    "\n",
    "    d_seeds_idsym = {}\n",
    "    for k,v in d_gene_sym.items():\n",
    "        for i in seeds_list:\n",
    "            if v == i:\n",
    "                d_seeds_idsym[k]=i\n",
    "    seed_list = list(d_seeds_idsym.keys())\n",
    "\n",
    "    FM_BP = pd.read_csv('input/Features_GO_BiolProc_Dataframe_human.csv', index_col=0)\n",
    "    bioproc_list = list(FM_BP.index)\n",
    "\n",
    "    # From obtained Gene lists select those to be mapped onto different shells \n",
    "    # select genes for shells \n",
    "    shell_one = variant_list\n",
    "    shell_two = seed_list\n",
    "    shell_three = [str(i) for i in bioproc_list]\n",
    "\n",
    "    small_rad = 1\n",
    "    mid_rad = 5\n",
    "    large_rad = 30\n",
    "    outershell_rad = 50\n",
    "\n",
    "    d_param = {}\n",
    "    for i in G.nodes():\n",
    "        if str(i) in variant_list:\n",
    "            d_param[str(i)] = small_rad\n",
    "        elif i in seed_list:\n",
    "            d_param[str(i)] = mid_rad \n",
    "        elif int(i) in bioproc_list:\n",
    "            d_param[str(i)] = large_rad \n",
    "        else:\n",
    "            d_param[str(i)] = outershell_rad\n",
    "\n",
    "    d_radius = d_param\n",
    "    genes_rad = variant_list + seed_list + bioproc_list\n",
    "\n",
    "    # check how many shells of spheres there will be :\n",
    "    print('Number of Spherical Shells:', len(set(d_param.values())))\n",
    "\n",
    "\n",
    "    # Specifying coloring based on genes on different shells \n",
    "    d_nodecol = d_clos \n",
    "    d_colors = color_nodes_from_dict(G, d_nodecol, palette = col_pal)\n",
    "\n",
    "    d_colors_spec_genes = {}\n",
    "    for k,v in d_colors.items():\n",
    "        if k in shell_one:\n",
    "            d_colors_spec_genes[k]='#8b0000' \n",
    "        elif k in shell_two:\n",
    "            d_colors_spec_genes[k]='#FF4500' \n",
    "        elif k in shell_three:\n",
    "            d_colors_spec_genes[k]=v\n",
    "        else: \n",
    "            d_colors_spec_genes[k]='#d3d3d3'\n",
    "\n",
    "    d_colors_spec_genes_sort = {key:d_colors_spec_genes[key] for key in G.nodes()}\n",
    "    colors = list(d_colors_spec_genes_sort.values())\n",
    "\n",
    "    \n",
    "elif r_feat == 'essentiality':\n",
    "\n",
    "    small_rad = 1\n",
    "    mid_rad = 10\n",
    "    outershell_rad = 20\n",
    "\n",
    "    d_ess_scores = {}\n",
    "    for i in G.nodes():\n",
    "        if i in essential_genes:\n",
    "            d_ess_scores[i] = small_rad\n",
    "        elif i in non_ess_genes:\n",
    "            d_ess_scores[i] = mid_rad\n",
    "        elif str(i) in notdefined_genes:\n",
    "            d_ess_scores[i] = outershell_rad\n",
    "\n",
    "    d_radius = d_ess_scores\n",
    "    genes_rad = list(G.nodes())\n",
    "\n",
    "    # check how many shells of spheres there will be :\n",
    "    print('Number of Spherical Shells:', len(set(d_param.values())))\n",
    "\n",
    "    # Specifying coloring based on genes on different shells \n",
    "    d_colors_spec_genes = {}\n",
    "    for k in G.nodes():\n",
    "        if k in essential_genes:\n",
    "            d_colors_spec_genes[k]='#0080bf' \n",
    "        elif k in non_ess_genes:\n",
    "            d_colors_spec_genes[k]='#95dcff' \n",
    "        else: \n",
    "            d_colors_spec_genes[k]='#d3d3d3'\n",
    "\n",
    "    d_colors_spec_genes_sort = {key:d_colors_spec_genes[key] for key in G.nodes()}\n",
    "    colors = list(d_colors_spec_genes_sort.values())\n",
    "    \n",
    "    \n",
    "else: \n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "standard-nursery",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### PLOT "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "driven-sword",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DM precalc used\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'output_plots/Geodesic/geodesic_r_r-autocore_local_human_light.html'"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ---------------------------------------------------------------------------\n",
    "#\n",
    "# SELECT a r - Parameter:\n",
    "#\n",
    "# d_radius > a dictionary with keys=G.nodes and values=any radius assigned to each node\n",
    "# ---------------------------------------------------------------------------\n",
    "\n",
    "posG_sphere = layout_geodesic(G, d_radius, n_neighbors=20, spread=1, min_dist=0.0)#, DM=DM)\n",
    "posG = posG_sphere\n",
    "\n",
    "umap_nodes = get_trace_nodes_3D(posG, l_features, colors, size3d, opacity_nodes)\n",
    "umap_edges = get_trace_edges_3D(G, posG, edge_color, opac = 0.05, linewidth = 0.5)\n",
    "data = [umap_nodes]#,umap_edges]\n",
    "\n",
    "path = 'output_plots/Geodesic/'\n",
    "schema = 'light' #'dark'\n",
    "fname = 'geodesic_r_'+r_feat+'_'+netlayout+'_'+organism+'_'+schema\n",
    "\n",
    "plot_3D(data,path,fname, 'dark')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:miniconda3-myenv_clip2]",
   "language": "python",
   "name": "conda-env-miniconda3-myenv_clip2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
